{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 本Script包含在Bacdive数据集和human gut&sputum数据集上进行引物设计和验证的代码\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "import math\n",
    "from collections import Counter\n",
    "\n",
    "# biopython utils\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from scipy.stats import wilcoxon\n",
    "from scipy.stats import chi2_contingency\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "\n",
    "# from adjustText import adjust_text\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. 首先处理silva数据库获得用于引物设计所必须的ref文件\n",
    "Model_data/Silva_ref_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 首先利用16sDeepSeg处理silva数据获得用于引物设计所必须的所有silva seqs的划分结果文件\n",
    "# 1106: 注意要把silva_id改成id才可以，之前是用silva_id_wrong跑的结果\n",
    "deepSeg_cmd = 'python 16sDeepSeg/module_output.py -c Model_data/16sDeepSeg_publicated_model/config.json -r Model_data/16sDeepSeg_publicated_model/model.pth --input /data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/data/SILVA_data/silva_seqs_id_only.csv --output data/SILVA_processed_data/SILVA_set_segmentation_16sDeepSeg.csv'\n",
    "os.system(deepSeg_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>silva_id</th>\n",
       "      <th>silva_id_wrong</th>\n",
       "      <th>16s_rna</th>\n",
       "      <th>lens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB001445.1.1538</td>\n",
       "      <td>AB001445.1.1538</td>\n",
       "      <td>AACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGC...</td>\n",
       "      <td>1538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KM209255.204.1909</td>\n",
       "      <td>KM209255.204.1909</td>\n",
       "      <td>AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...</td>\n",
       "      <td>1706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HL281554.1.1313</td>\n",
       "      <td>HL281554.1.1313</td>\n",
       "      <td>GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAGTGG...</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB002515.1.1332</td>\n",
       "      <td>AB002515.1.1332</td>\n",
       "      <td>GCCTAATACATGCAAGTTGACGACAGATGATACGTAGCTTGCTACA...</td>\n",
       "      <td>1332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB002523.1.1496</td>\n",
       "      <td>AB002523.1.1496</td>\n",
       "      <td>TCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAATACATGCAAGT...</td>\n",
       "      <td>1496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            silva_id     silva_id_wrong  \\\n",
       "0    AB001445.1.1538    AB001445.1.1538   \n",
       "1  KM209255.204.1909  KM209255.204.1909   \n",
       "2    HL281554.1.1313    HL281554.1.1313   \n",
       "3    AB002515.1.1332    AB002515.1.1332   \n",
       "4    AB002523.1.1496    AB002523.1.1496   \n",
       "\n",
       "                                             16s_rna  lens  \n",
       "0  AACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGC...  1538  \n",
       "1  AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...  1706  \n",
       "2  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGAGTGG...  1313  \n",
       "3  GCCTAATACATGCAAGTTGACGACAGATGATACGTAGCTTGCTACA...  1332  \n",
       "4  TCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAATACATGCAAGT...  1496  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_silva = pd.read_csv('data/SILVA_data/silva_seqs_id_only.csv', )\n",
    "df_silva.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 11/06/2023 19:12:28\n",
      "New DB name:   /data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/data/ecoli_db/Ecoli_ref_1550\n",
      "New DB title:  /data1/hyzhang/Projects/16sDeepSeg_summary/Evi_specific_primers_database/Primer_evaluation_fromMetagenomeData/data/Ecoli_ref.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1 sequences in 0.005512 seconds.\n"
     ]
    }
   ],
   "source": [
    "# makeblastdb: 现在用的是忘了从哪里搞得，比Ecoli K12的NCBI序列更长？后续换一下？\n",
    "!makeblastdb -in /data1/hyzhang/Projects/16sDeepSeg_summary/Evi_specific_primers_database/Primer_evaluation_fromMetagenomeData/data/Ecoli_ref.fasta -input_type fasta -dbtype nucl -out data/ecoli_db/Ecoli_ref_1550"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['query_acc.ver', 'subject_acc.ver', '%_identity', 'alignment_length', 'mismatches', 'gap_opens', 'q._start', 'q._end', 's._start', 's._end', 'evalue', 'bit_score']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 431575/431575 [7:32:48<00:00, 15.89it/s]   \n"
     ]
    }
   ],
   "source": [
    "## 计算所有silva seqs中包含了完整的ecoli 8bp-1492bp的序列\n",
    "## 这个script用来构建包含全长序列的silva数据库，从431388的全长序列出发，就跟mothur无关了\n",
    "def blast_cmd(query_primer=\"\", db=\"\", out_path = ''):\n",
    "    cmd = f\"blastn -db {db} -query \"\n",
    "    cmd += query_primer\n",
    "    cmd += \" -outfmt 7 \"\n",
    "    cmd += \"-out \"\n",
    "    cmd += out_path\n",
    "    cmd += \" -task blastn-short -word_size 4 -evalue 1 -max_target_seqs 1000000\"\n",
    "    \n",
    "    os.system(cmd)\n",
    "\n",
    "# primer blast results\n",
    "def parse_blastTxt(blast_txt = '/data3/hyzhang/ont/16s_RNA_seg/res/blast_res/query.txt'):\n",
    "    with open(blast_txt, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    index_line = lines[3]\n",
    "    index_line = index_line.split(': ')[-1]\n",
    "    index_line = index_line.split(',')\n",
    "    index_line = [x.strip(' ') for x in index_line]\n",
    "    index_line = ['_'.join(x.split()) for x in index_line]\n",
    "    print(index_line)\n",
    "\n",
    "    lines = [x.strip('\\n').split('\\t') for x in lines if not x.startswith('#')]\n",
    "    \n",
    "    blast_df = pd.DataFrame(lines)\n",
    "    blast_df.columns = index_line\n",
    "    \n",
    "    # os.remove(blast_txt)\n",
    "    return blast_df\n",
    "\n",
    "\n",
    "# 以下是序列比对和操作过程\n",
    "ecoli_fromPos_cut = 8\n",
    "ecoli_endPos_cut = 1500\n",
    "\n",
    "def filter_seqs(silva_ori_file = 'data/SILVA_data/silva_seqs_id_only.csv'):\n",
    "    # read csv info file containing all the taxa and seq ids.\n",
    "    silva_database_seqs = pd.read_csv(silva_ori_file)\n",
    "    # write a fasta file\n",
    "    records = []\n",
    "    for i in range(silva_database_seqs.shape[0]):\n",
    "        seq_now = silva_database_seqs['16s_rna'][i]\n",
    "        seq_now = Seq(seq_now)\n",
    "        id_now = silva_database_seqs['silva_id'][i]\n",
    "        \n",
    "        rec = SeqRecord(seq=seq_now, id=id_now, description= '')\n",
    "        records.append(rec)\n",
    "\n",
    "    with open('data/SILVA_processed_data/silva_seqs_all_1106.fasta', 'w') as f:\n",
    "        SeqIO.write(records, f, 'fasta')\n",
    "        \n",
    "    # blast every seqs to Ecoli ref seq\n",
    "    silva_seqs = 'data/SILVA_processed_data/silva_seqs_all_1106.fasta'\n",
    "    out_path = silva_seqs.replace(\".fasta\", \"_alignToEcoli.fasta\")\n",
    "    db = 'data/ecoli_db/Ecoli_ref_1550'\n",
    "\n",
    "    blast_cmd(query_primer=silva_seqs, out_path=out_path, db=db)\n",
    "\n",
    "    # parse res of the blast\n",
    "    from tqdm import tqdm\n",
    "    blast_df = parse_blastTxt(out_path)\n",
    "    need_to_be_int = ['s._start', 's._end', 'q._start', 'q._end', 'alignment_length']\n",
    "    for ind in need_to_be_int:\n",
    "        blast_df[ind] = blast_df[ind].apply(int)\n",
    "    blast_df['evalue'] = blast_df['evalue'].apply(float)\n",
    "    blast_res_df = []\n",
    "    for seq_id, seq_df in tqdm(blast_df.groupby('query_acc.ver')):\n",
    "        res_t = {'seq_id':seq_id}\n",
    "        seq_df.sort_values(by='s._start', ascending=True, inplace=True)\n",
    "        seq_df = seq_df[(seq_df['alignment_length'] >= 15) & (seq_df['evalue'] < 0.5)]\n",
    "        seq_df.reset_index(inplace=True, drop=True)\n",
    "        \n",
    "        silva_info = silva_database_seqs[silva_database_seqs['silva_id'] == seq_id]\n",
    "        seq_lens = int(silva_info.reset_index()['lens'][0]) # full lens including the 8f and 1492r region\n",
    "        \n",
    "        # get info \n",
    "        qs1, qe1, ss1, se1 = list(seq_df.loc[0, ['q._start', 'q._end', 's._start', 's._end']]) \n",
    "        qs2, qe2, ss2, se2 = list(seq_df.loc[seq_df.shape[0] - 1, ['q._start', 'q._end', 's._start', 's._end']])\n",
    "        \n",
    "        start_bp_correspondingSiteInEcoli = ss1 - qs1\n",
    "        end_bp_correspondingSiteInEcoli = seq_lens - qe2 + se2\n",
    "        res_t['start_bp_correspondingSiteInEcoli'] = start_bp_correspondingSiteInEcoli\n",
    "        res_t['end_bp_correspondingSiteInEcoli'] = end_bp_correspondingSiteInEcoli\n",
    "        blast_res_df.append(res_t)\n",
    "        \n",
    "    blast_res_df = pd.DataFrame(blast_res_df)\n",
    "    \n",
    "    # chose the cut-off value.\n",
    "    silva_with_complete_8f_and_1492r = blast_res_df[(blast_res_df['start_bp_correspondingSiteInEcoli'] <= ecoli_fromPos_cut) & (blast_res_df['end_bp_correspondingSiteInEcoli'] >= ecoli_endPos_cut)]\n",
    "    silva_with_complete_8f_and_1492r.reset_index(inplace=True, drop=True)\n",
    "    silva_with_complete_8f_and_1492r.to_csv(f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r.csv', index=False)\n",
    "\n",
    "def merge_with_metadata(filter_csv = f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r.csv'):\n",
    "    ori_silva_meta = pd.read_csv('data/SILVA_data/silva_metainfo_clear.csv')\n",
    "    ori_silva_idSeqs = pd.read_csv('data/SILVA_data/silva_seqs_id_only.csv')\n",
    "    filtered_seqs = pd.read_csv(filter_csv)\n",
    "    # 按照silva id 进行merge，然后保留wrong id进行后续primer design\n",
    "    filter_silva_meta = ori_silva_meta[ori_silva_meta['silva_id'].isin(list(filtered_seqs['seq_id']))]\n",
    "    filter_silva_idSeqs = ori_silva_idSeqs[ori_silva_idSeqs['silva_id'].isin(list(filtered_seqs['seq_id']))]\n",
    "    filter_silva_meta.reset_index(inplace = True, drop = True)\n",
    "    filter_silva_idSeqs.reset_index(inplace = True, drop = True)\n",
    "    \n",
    "    filter_silva_meta.to_csv(f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r_TaxaMeta.csv', index=False)\n",
    "    filter_silva_idSeqs.to_csv(f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r_IdSeqs.csv', index=False)\n",
    "    \n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    filter_seqs()\n",
    "    merge_with_metadata()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "156813\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>silva_id</th>\n",
       "      <th>silva_id_wrong</th>\n",
       "      <th>16s_rna</th>\n",
       "      <th>taxa_info_summary</th>\n",
       "      <th>ori_rank_tree_num</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB001445.1.1538</td>\n",
       "      <td>AB001445.1.1538</td>\n",
       "      <td>AACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGC...</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Ps...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas amygdali pv. morsprunorum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KM209255.204.1909</td>\n",
       "      <td>KM209255.204.1909</td>\n",
       "      <td>AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Pectobacteriaceae</td>\n",
       "      <td>Dickeya</td>\n",
       "      <td>Dickeya phage phiDP10.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JN088732.1.1500</td>\n",
       "      <td>JN088732.1.1500</td>\n",
       "      <td>AGAGTTTGATCMTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;Ps...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Porticoccaceae</td>\n",
       "      <td>Porticoccus</td>\n",
       "      <td>Porticoccus hydrocarbonoclasticus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JN167515.1.1446</td>\n",
       "      <td>JN167515.1.1446</td>\n",
       "      <td>AGAGTTTGATCCTGGCTCAGAACGAACGCTGGCGGCAGGCCTAACA...</td>\n",
       "      <td>Bacteria;Proteobacteria;Alphaproteobacteria;Rh...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhizobiales</td>\n",
       "      <td>Stappiaceae</td>\n",
       "      <td>Pseudovibrio</td>\n",
       "      <td>Pseudovibrio axinellae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG726039.1.1478</td>\n",
       "      <td>HG726039.1.1478</td>\n",
       "      <td>AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAACA...</td>\n",
       "      <td>Bacteria;Firmicutes;Clostridia;Clostridiales;C...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Clostridiaceae</td>\n",
       "      <td>Clostridium sensu stricto 1</td>\n",
       "      <td>Clostridium saudiense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156808</th>\n",
       "      <td>CFFF01000203.442.2194</td>\n",
       "      <td>CFFF01000203.442.2194</td>\n",
       "      <td>ACAGGCTTATCTCCCCCAAGAGTTCACATCGACGGGGAGGTTTGGC...</td>\n",
       "      <td>Bacteria;Firmicutes;Bacilli;Lactobacillales;St...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156809</th>\n",
       "      <td>CXYW01000009.360.1734</td>\n",
       "      <td>CXYW01000009.360.1734</td>\n",
       "      <td>GAGAGTTTGATCCTGGCTCAGGATGAACGCTGGCGGCGTGCTTAAC...</td>\n",
       "      <td>Bacteria;Firmicutes;Clostridia;Lachnospirales;...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Epulopiscium</td>\n",
       "      <td>Lachnospiraceae bacterium mt14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156810</th>\n",
       "      <td>CXWL01024289.1436.2731</td>\n",
       "      <td>CXWL01024289.1436.2731</td>\n",
       "      <td>AGAGTTTGATCCTAGCTCAGGGTGAATGCTGGCGGCGTGGATAAGT...</td>\n",
       "      <td>Bacteria;Patescibacteria;Parcubacteria;Candida...</td>\n",
       "      <td>5</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Patescibacteria</td>\n",
       "      <td>Parcubacteria</td>\n",
       "      <td>Candidatus Kaiserbacteria</td>\n",
       "      <td>groundwater metagenome</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156811</th>\n",
       "      <td>MVBC01000034.3.1520</td>\n",
       "      <td>MVBC01000034.3.1520</td>\n",
       "      <td>TTACAATGGAGAGTTTGATCCTGGCTCAGGATGAACGCTAGCGGCA...</td>\n",
       "      <td>Bacteria;Bacteroidota;Bacteroidia;Cytophagales...</td>\n",
       "      <td>7</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Cytophagales</td>\n",
       "      <td>Hymenobacteraceae</td>\n",
       "      <td>Hymenobacter</td>\n",
       "      <td>Hymenobacter sp. CRA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156812</th>\n",
       "      <td>JQHZ01065714.557.1975</td>\n",
       "      <td>JQHZ01065714.557.1975</td>\n",
       "      <td>AGAGTTCGATCCTGGCTCAGGATAAACGCTGGCAGCATGCTTAACA...</td>\n",
       "      <td>Bacteria;Cyanobacteria;Cyanobacteriia;Chloropl...</td>\n",
       "      <td>5</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Catharanthus roseus (Madagascar periwinkle)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156813 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      silva_id          silva_id_wrong  \\\n",
       "0              AB001445.1.1538         AB001445.1.1538   \n",
       "1            KM209255.204.1909       KM209255.204.1909   \n",
       "2              JN088732.1.1500         JN088732.1.1500   \n",
       "3              JN167515.1.1446         JN167515.1.1446   \n",
       "4              HG726039.1.1478         HG726039.1.1478   \n",
       "...                        ...                     ...   \n",
       "156808   CFFF01000203.442.2194   CFFF01000203.442.2194   \n",
       "156809   CXYW01000009.360.1734   CXYW01000009.360.1734   \n",
       "156810  CXWL01024289.1436.2731  CXWL01024289.1436.2731   \n",
       "156811     MVBC01000034.3.1520     MVBC01000034.3.1520   \n",
       "156812   JQHZ01065714.557.1975   JQHZ01065714.557.1975   \n",
       "\n",
       "                                                  16s_rna  \\\n",
       "0       AACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGC...   \n",
       "1       AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...   \n",
       "2       AGAGTTTGATCMTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACA...   \n",
       "3       AGAGTTTGATCCTGGCTCAGAACGAACGCTGGCGGCAGGCCTAACA...   \n",
       "4       AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAACA...   \n",
       "...                                                   ...   \n",
       "156808  ACAGGCTTATCTCCCCCAAGAGTTCACATCGACGGGGAGGTTTGGC...   \n",
       "156809  GAGAGTTTGATCCTGGCTCAGGATGAACGCTGGCGGCGTGCTTAAC...   \n",
       "156810  AGAGTTTGATCCTAGCTCAGGGTGAATGCTGGCGGCGTGGATAAGT...   \n",
       "156811  TTACAATGGAGAGTTTGATCCTGGCTCAGGATGAACGCTAGCGGCA...   \n",
       "156812  AGAGTTCGATCCTGGCTCAGGATAAACGCTGGCAGCATGCTTAACA...   \n",
       "\n",
       "                                        taxa_info_summary  ori_rank_tree_num  \\\n",
       "0       Bacteria;Proteobacteria;Gammaproteobacteria;Ps...                  7   \n",
       "1       Bacteria;Proteobacteria;Gammaproteobacteria;En...                  7   \n",
       "2       Bacteria;Proteobacteria;Gammaproteobacteria;Ps...                  7   \n",
       "3       Bacteria;Proteobacteria;Alphaproteobacteria;Rh...                  7   \n",
       "4       Bacteria;Firmicutes;Clostridia;Clostridiales;C...                  7   \n",
       "...                                                   ...                ...   \n",
       "156808  Bacteria;Firmicutes;Bacilli;Lactobacillales;St...                  7   \n",
       "156809  Bacteria;Firmicutes;Clostridia;Lachnospirales;...                  7   \n",
       "156810  Bacteria;Patescibacteria;Parcubacteria;Candida...                  5   \n",
       "156811  Bacteria;Bacteroidota;Bacteroidia;Cytophagales...                  7   \n",
       "156812  Bacteria;Cyanobacteria;Cyanobacteriia;Chloropl...                  5   \n",
       "\n",
       "         Kingdom           Phylum                Class  \\\n",
       "0       Bacteria   Proteobacteria  Gammaproteobacteria   \n",
       "1       Bacteria   Proteobacteria  Gammaproteobacteria   \n",
       "2       Bacteria   Proteobacteria  Gammaproteobacteria   \n",
       "3       Bacteria   Proteobacteria  Alphaproteobacteria   \n",
       "4       Bacteria       Firmicutes           Clostridia   \n",
       "...          ...              ...                  ...   \n",
       "156808  Bacteria       Firmicutes              Bacilli   \n",
       "156809  Bacteria       Firmicutes           Clostridia   \n",
       "156810  Bacteria  Patescibacteria        Parcubacteria   \n",
       "156811  Bacteria     Bacteroidota          Bacteroidia   \n",
       "156812  Bacteria    Cyanobacteria       Cyanobacteriia   \n",
       "\n",
       "                            Order  \\\n",
       "0                 Pseudomonadales   \n",
       "1                Enterobacterales   \n",
       "2                 Pseudomonadales   \n",
       "3                     Rhizobiales   \n",
       "4                   Clostridiales   \n",
       "...                           ...   \n",
       "156808            Lactobacillales   \n",
       "156809             Lachnospirales   \n",
       "156810  Candidatus Kaiserbacteria   \n",
       "156811               Cytophagales   \n",
       "156812                Chloroplast   \n",
       "\n",
       "                                             Family  \\\n",
       "0                                  Pseudomonadaceae   \n",
       "1                                 Pectobacteriaceae   \n",
       "2                                    Porticoccaceae   \n",
       "3                                       Stappiaceae   \n",
       "4                                    Clostridiaceae   \n",
       "...                                             ...   \n",
       "156808                             Streptococcaceae   \n",
       "156809                              Lachnospiraceae   \n",
       "156810                       groundwater metagenome   \n",
       "156811                            Hymenobacteraceae   \n",
       "156812  Catharanthus roseus (Madagascar periwinkle)   \n",
       "\n",
       "                              Genus                                Species  \n",
       "0                       Pseudomonas  Pseudomonas amygdali pv. morsprunorum  \n",
       "1                           Dickeya                Dickeya phage phiDP10.3  \n",
       "2                       Porticoccus      Porticoccus hydrocarbonoclasticus  \n",
       "3                      Pseudovibrio                 Pseudovibrio axinellae  \n",
       "4       Clostridium sensu stricto 1                  Clostridium saudiense  \n",
       "...                             ...                                    ...  \n",
       "156808                Streptococcus               Streptococcus pneumoniae  \n",
       "156809                 Epulopiscium         Lachnospiraceae bacterium mt14  \n",
       "156810                      unknown                                unknown  \n",
       "156811                 Hymenobacter                  Hymenobacter sp. CRA2  \n",
       "156812                      unknown                                unknown  \n",
       "\n",
       "[156813 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 生成后续用于in-silico验证的关键文件\n",
    "\n",
    "# 1106: 拼接一下目前筛选出来的14w的seqs们\n",
    "idseq_df = pd.read_csv(f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r_IdSeqs.csv')\n",
    "idseq_df.drop(columns=['lens'], axis=1, inplace=True)\n",
    "meta_df = pd.read_csv(f'data/SILVA_processed_data/silva_with_complete_{ecoli_fromPos_cut}f_{ecoli_endPos_cut}r_TaxaMeta.csv')\n",
    "\n",
    "id_seq_taxa_df = pd.merge(idseq_df, meta_df, on=['silva_id', 'silva_id_wrong'], how='inner')\n",
    "filter_seqs_df = id_seq_taxa_df\n",
    "\n",
    "# write the fasta file to the corresponding fasta.\n",
    "silva_nr_filtered_fasta = []\n",
    "for seq, id_, des in zip(list(filter_seqs_df['16s_rna']), list(filter_seqs_df['silva_id']), list(filter_seqs_df['taxa_info_summary'])):\n",
    "    seq_now = Seq(seq)\n",
    "    rec = SeqRecord(seq=seq_now, id=id_, description=des)\n",
    "    \n",
    "    if 'Mitochondria' in des:\n",
    "        continue\n",
    "    silva_nr_filtered_fasta.append(rec)\n",
    "# SeqIO.write(silva_nr_filtered_fasta, 'data/SILVA_processed_data/InsilicoPCR_for_PrimerMatch/silva_seqs_157515_10f_1500r.fasta', 'fasta')\n",
    "\n",
    "\n",
    "# write the metainfo containing taxa_info and rna seqs\n",
    "filter_seqs_notMito_df = filter_seqs_df[filter_seqs_df['taxa_info_summary'].apply(lambda x: 'Mitochondria' not in x)]\n",
    "filter_seqs_notMito_df.reset_index(inplace=True, drop=True)\n",
    "# filter_seqs_notMito_df = pd.DataFrame(filter_seqs_notMito_df, columns=['silva_id_ori_fromSilva', 'taxa_info', '16s_rna'])\n",
    "\n",
    "filter_seqs_notMito_df = filter_seqs_notMito_df[filter_seqs_notMito_df['Genus'].apply(lambda x: len(x) > 1)]\n",
    "filter_seqs_notMito_df.reset_index(inplace = True, drop = True)\n",
    "print(filter_seqs_notMito_df.shape[0])\n",
    "filter_seqs_notMito_df.to_csv('data/SILVA_processed_data/InsilicoPCR_for_PrimerMatch/silva_metaInfo_157515_10f_1500r.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Bacdive数据库中多种类型环境的in-silico验证结果\n",
    "\n",
    "首先清洗bacdive数据库"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original dataset shape: (126545, 6)\n"
     ]
    }
   ],
   "source": [
    "## 首先bacdive数据库中的genus id和silva id对应\n",
    "# original data and data root\n",
    "bacdive_ori_data = 'data/Bacdive/isolation_sources_bacdive_2022-10-07.csv'\n",
    "bacdive_root = 'data/Bacdive'\n",
    "\n",
    "# create a cross table\n",
    "cols = ['ID', 'Species', 'Isolation source', 'Country', 'Continent', 'Category 3']\n",
    "bacdive_ori_df = pd.read_csv(bacdive_ori_data)\n",
    "bacdive_ori_df = pd.DataFrame(bacdive_ori_df, columns=cols)\n",
    "print('Original dataset shape:', bacdive_ori_df.shape)\n",
    "\n",
    "# fill id and species na\n",
    "bacdive_fillID_df = bacdive_ori_df.copy()\n",
    "bacdive_fillID_df['ID'] = bacdive_fillID_df['ID'].fillna(method='ffill')\n",
    "bacdive_fillID_df['Species'] = bacdive_fillID_df['Species'].fillna(method='ffill')\n",
    "\n",
    "# create multi-label dataframe\n",
    "bacdive_env_labels_Species_df = pd.crosstab(bacdive_fillID_df['Species'], bacdive_fillID_df['Category 3'])\n",
    "bacdive_env_labels_Species_df = pd.DataFrame(bacdive_env_labels_Species_df)\n",
    "bacdive_env_labels_Species_df.columns = list(bacdive_env_labels_Species_df.columns)\n",
    "bacdive_env_labels_Species_df.to_csv(os.path.join(bacdive_root, 'bacdive_micro_Species_labels_1030.tsv'), sep = '\\t')\n",
    "\n",
    "# convert the species name to species ID\n",
    "# !Rscript convert_taxid_Bacdive.r # output file bacdive_micro_Species_labels_taxid.tsv\n",
    "\n",
    "# get all the species id for bacdive taxa\n",
    "bacdive_id = 'data/Bacdive/bacdive_micro_Species_labels_taxid.tsv'\n",
    "bacdive_taxid = pd.read_csv(bacdive_id)\n",
    "bacdive_taxid = list(bacdive_taxid['Species'])\n",
    "bacdive_taxid = [str(int(x)) if (not math.isnan(x)) else '-1' for x in bacdive_taxid]\n",
    "bacdive_taxid = list(set(bacdive_taxid))\n",
    "\n",
    "bacdive_taxid = {'bacdive_taxid': bacdive_taxid}\n",
    "bacdive_taxid = pd.DataFrame(bacdive_taxid)\n",
    "bacdive_taxid.to_csv(f'{bacdive_root}/bacdive_all_species_taxid.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "使用外源的R代码根据species id获取其各层级的物种id，bacdive_all_taxa.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original dataset shape: (126545, 6)\n",
      "Only bacteria shape: (118002, 10)\n",
      "With genus name shape: (117680, 10)\n",
      "         ID                      Species  \\\n",
      "0  159652.0  Abditibacterium utsteinense   \n",
      "1  159652.0  Abditibacterium utsteinense   \n",
      "2     219.0        Abiotrophia defectiva   \n",
      "3     219.0        Abiotrophia defectiva   \n",
      "4     219.0        Abiotrophia defectiva   \n",
      "\n",
      "                                    Isolation source     Country  \\\n",
      "0  Top surface sample consisting of weathered gra...  Antarctica   \n",
      "1                                                NaN         NaN   \n",
      "2            blood of bacterial endocarditis patient      France   \n",
      "3                                                NaN         NaN   \n",
      "4                                                NaN         NaN   \n",
      "\n",
      "               Continent Category 3               bacdive_spe_nm   genus_id  \\\n",
      "0  Australia and Oceania  #Geologic  Abditibacterium utsteinense  2109263.0   \n",
      "1                    NaN    #Alpine  Abditibacterium utsteinense  2109263.0   \n",
      "2                 Europe     #Blood        Abiotrophia defectiva    46123.0   \n",
      "3                    NaN        NaN        Abiotrophia defectiva    46123.0   \n",
      "4                    NaN        NaN        Abiotrophia defectiva    46123.0   \n",
      "\n",
      "        genus_name kingdom_name  \n",
      "0  Abditibacterium     Bacteria  \n",
      "1  Abditibacterium     Bacteria  \n",
      "2      Abiotrophia     Bacteria  \n",
      "3      Abiotrophia     Bacteria  \n",
      "4      Abiotrophia     Bacteria  \n"
     ]
    }
   ],
   "source": [
    "# 有了NCBI获取的species id之后合并table，保存bacdive所有species的taxonomic metainfo\n",
    "bacdive_spe = pd.read_csv(f'{bacdive_root}/bacdive_micro_Species_labels.tsv', sep='\\t')\n",
    "bacdive_spe = list(bacdive_spe['Species'])\n",
    "\n",
    "bacdive_taxid = pd.read_csv(f'{bacdive_root}/bacdive_micro_Species_labels_taxid.tsv')\n",
    "bacdive_taxid = list(bacdive_taxid['Species'])\n",
    "bacdive_taxid = [str(int(x)) if (not math.isnan(x)) else '-1' for x in bacdive_taxid]\n",
    "\n",
    "bacdive_metainfo = {'bacdive_spe_nm': bacdive_spe, 'taxaid': bacdive_taxid}\n",
    "bacdive_metainfo = pd.DataFrame(bacdive_metainfo)\n",
    "\n",
    "jinyuan_parsed_speids = f'{bacdive_root}/bacdive_all_taxa.csv'\n",
    "jinyuan_parsed_df = pd.read_csv(jinyuan_parsed_speids)\n",
    "jinyuan_parsed_df['taxaid'] = jinyuan_parsed_df['taxaid'].apply(str)\n",
    "\n",
    "bacdive_metainfo = pd.merge(bacdive_metainfo, jinyuan_parsed_df, on = 'taxaid', how = 'left')\n",
    "for col in ['taxaid', 'genus_id', 'species_id']:\n",
    "    bacdive_metainfo[col].fillna(-1, inplace=True)\n",
    "    bacdive_metainfo[col] = bacdive_metainfo[col].apply(lambda x: str(int(x)))\n",
    "bacdive_metainfo.to_csv(f'{bacdive_root}/bacdive_metainfo.csv', index=False)\n",
    "\n",
    "# read bacdive_metainfo.csv\n",
    "bacdive_metainfo = pd.read_csv(f'{bacdive_root}/bacdive_metainfo.csv')\n",
    "bacdive_metainfo_only_genus = pd.DataFrame(bacdive_metainfo, columns=['bacdive_spe_nm', 'genus_id', 'genus_name', 'kingdom_name'])\n",
    "bacdive_metainfo_only_genus.head()\n",
    "\n",
    "cols = ['ID', 'Species', 'Isolation source', 'Country', 'Continent', 'Category 3']\n",
    "bacdive_ori_df = pd.read_csv(bacdive_ori_data)\n",
    "bacdive_ori_df = pd.DataFrame(bacdive_ori_df, columns=cols)\n",
    "print('Original dataset shape:', bacdive_ori_df.shape)\n",
    "\n",
    "# fill id and species na\n",
    "bacdive_fillID_df = bacdive_ori_df.copy()\n",
    "bacdive_fillID_df['ID'] = bacdive_fillID_df['ID'].fillna(method='ffill')\n",
    "bacdive_fillID_df['Species'] = bacdive_fillID_df['Species'].fillna(method='ffill')\n",
    "\n",
    "# simple way to translate species to genus\n",
    "bacdive_fillID_df = pd.merge(bacdive_fillID_df, bacdive_metainfo_only_genus, left_on='Species', right_on='bacdive_spe_nm', how='left')\n",
    "bacdive_fillID_df = bacdive_fillID_df[bacdive_fillID_df['kingdom_name'] == 'Bacteria']\n",
    "print('Only bacteria shape:', bacdive_fillID_df.shape)\n",
    "bacdive_fillID_df = bacdive_fillID_df[(~bacdive_fillID_df['genus_name'].isnull()) & (bacdive_fillID_df['genus_name'] != '-1')]\n",
    "print('With genus name shape:', bacdive_fillID_df.shape)\n",
    "bacdive_fillID_df.reset_index(inplace=True,drop=True)\n",
    "print(bacdive_fillID_df.head())\n",
    "\n",
    "# to genus level\n",
    "# create multi-label dataframe\n",
    "bacdive_env_labels_Species_df = pd.crosstab(bacdive_fillID_df['genus_name'], bacdive_fillID_df['Category 3'])\n",
    "bacdive_env_labels_Species_df = pd.DataFrame(bacdive_env_labels_Species_df)\n",
    "bacdive_env_labels_Species_df.columns = list(bacdive_env_labels_Species_df.columns)\n",
    "\n",
    "# create environment info table\n",
    "bacdive_env_labels_Species_df[bacdive_env_labels_Species_df >= 1] = 1\n",
    "bacdive_env_micro_spe = bacdive_env_labels_Species_df.sum(axis=0).sort_values(ascending=False)\n",
    "\n",
    "bacdive_env_micro_df = pd.DataFrame({'Genus_num': bacdive_env_micro_spe})\n",
    "bacdive_env_micro_df['Information'] = [''] * bacdive_env_micro_df.shape[0]\n",
    "bacdive_env_micro_df['Micro_spe'] = [''] * bacdive_env_micro_df.shape[0]\n",
    "envs = list(bacdive_env_micro_df.index)\n",
    "\n",
    "for env in envs:\n",
    "    bacdive_env_labels_Species_df_t = bacdive_env_labels_Species_df[bacdive_env_labels_Species_df[env] >= 1]\n",
    "    env_micro_ls = list(bacdive_env_labels_Species_df_t.index)\n",
    "    bacdive_env_micro_df.loc[env, 'Micro_spe'] = str(env_micro_ls)\n",
    "    \n",
    "    if len(env_micro_ls) != bacdive_env_micro_df.loc[env, 'Genus_num']:\n",
    "        print('aa')\n",
    "\n",
    "bacdive_env_micro_df.to_csv(os.path.join(bacdive_root, 'Env_micro_list_genus.tsv'), sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "对bacdive所有环境设计引物&in-silico PCR验证&挑选最优引物输出"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/285 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total set has 157404 seqs.\n",
      "Environment has 815 genus.\n",
      "Target v1v3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/9 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maybe unusual genus name: candidatus protochlamydia.\n",
      "Maybe unusual genus name: candidatus solibacter.\n",
      "Maybe unusual genus name: halalkalibacterium (ex joshi et al. 2022).\n",
      "Have -1. 1 -1\n",
      "Get 13510 records.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "muscle 5.1.linux64 [12f0e2]  1056Gb RAM, 192 cores\n",
      "Built Jan 13 2022 23:17:13\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 13510 seqs, length avg 99 max 99\n",
      "\n",
      "00:00 10Mb    100.0% Derep 8131 uniques, 5378 dupes\n",
      "00:00 12Mb   CPU has 192 cores, running 32 threads                   \n",
      "01:44 15Mb    100.0% UCLUST 8131 seqs EE<0.01, 3498 centroids, 4632 members\n",
      "03:06 2.4Gb   100.0% UCLUST 3498 seqs EE<0.30, 125 centroids, 3372 members \n",
      "03:06 2.4Gb   100.0% Make cluster MFAs                                    \n",
      "125 clusters pass 1                   \n",
      "03:27 2.4Gb   100.0% UCLUST 781 seqs EE<0.10, 101 centroids, 679 members\n",
      "03:27 2.4Gb   100.0% Make cluster MFAs                                  \n",
      "03:51 2.4Gb   100.0% UCLUST 761 seqs EE<0.10, 124 centroids, 636 members\n",
      "03:51 2.4Gb   100.0% Make cluster MFAs                                  \n",
      "348 clusters pass 2                   \n",
      "03:51 2.4Gb  \n",
      "03:51 2.4Gb  Align cluster 1 / 348 (61 seqs)\n",
      "03:51 2.4Gb  \n",
      "03:57 2.4Gb   100.0% Calc posteriors\n",
      "03:57 2.4Gb   100.0% Consistency (1/2)\n",
      "03:57 2.4Gb   100.0% Consistency (2/2)\n",
      "03:57 2.4Gb   100.0% UPGMA5           \n",
      "03:57 2.4Gb   100.0% Refining\n",
      "03:57 2.4Gb                  \n",
      "03:57 2.4Gb  Align cluster 2 / 348 (64 seqs)\n",
      "03:57 2.4Gb  \n",
      "03:58 2.4Gb   100.0% Calc posteriors\n",
      "03:58 2.4Gb   100.0% Consistency (1/2)\n",
      "03:59 2.4Gb   100.0% Consistency (2/2)\n",
      "03:59 2.4Gb   100.0% UPGMA5           \n",
      "03:59 2.4Gb   100.0% Refining\n",
      "03:59 2.4Gb                  \n",
      "03:59 2.4Gb  Align cluster 3 / 348 (3 seqs)\n",
      "03:59 2.4Gb  \n",
      "03:59 2.4Gb   100.0% Calc posteriors\n",
      "03:59 2.4Gb   100.0% Consistency (1/2)\n",
      "03:59 2.4Gb   100.0% Consistency (2/2)\n",
      "03:59 2.4Gb   100.0% UPGMA5           \n",
      "03:59 2.4Gb   100.0% Refining\n",
      "03:59 2.4Gb                  \n",
      "03:59 2.4Gb  Align cluster 4 / 348 (4 seqs)\n",
      "03:59 2.4Gb  \n",
      "03:59 2.4Gb   100.0% Calc posteriors\n",
      "03:59 2.4Gb   100.0% Consistency (1/2)\n",
      "03:59 2.4Gb   100.0% Consistency (2/2)\n",
      "03:59 2.4Gb   100.0% UPGMA5           \n",
      "03:59 2.4Gb   100.0% Refining\n",
      "03:59 2.4Gb                  \n",
      "03:59 2.4Gb  Align cluster 5 / 348 (28 seqs)\n",
      "03:59 2.4Gb  \n",
      "03:59 2.4Gb   100.0% Calc posteriorsalc posteriors\n",
      "03:59 2.4Gb   100.0% Consistency (1/2)\n",
      "03:59 2.4Gb   100.0% Consistency (2/2)\n",
      "03:59 2.4Gb   100.0% UPGMA5           \n",
      "03:59 2.4Gb   100.0% Refining\n",
      "03:59 2.4Gb                  \n",
      "03:59 2.4Gb  Align cluster 6 / 348 (207 seqs)\n",
      "03:59 2.4Gb  \n",
      "04:26 2.4Gb   100.0% Calc posteriors \n",
      "04:31 2.4Gb   100.0% Consistency (1/2) \n",
      "04:33 2.4Gb   100.0% Consistency (2/2) \n",
      "04:33 2.4Gb   100.0% UPGMA5           \n",
      "04:36 2.4Gb   100.0% Refining\n",
      "04:36 2.4Gb                  \n",
      "04:36 2.4Gb  Align cluster 7 / 348 (3 seqs)\n",
      "04:36 2.4Gb  \n",
      "04:36 2.4Gb   100.0% Calc posteriors\n",
      "04:36 2.4Gb   100.0% Consistency (1/2)\n",
      "04:36 2.4Gb   100.0% Consistency (2/2)\n",
      "04:36 2.4Gb   100.0% UPGMA5           \n",
      "04:36 2.4Gb   100.0% Refining\n",
      "04:36 2.4Gb                  \n",
      "04:36 2.4Gb  Align cluster 8 / 348 (83 seqs)\n",
      "04:36 2.4Gb  \n",
      "04:37 2.4Gb   100.0% Calc posteriors\n",
      "04:37 2.4Gb   100.0% Consistency (1/2)\n",
      "04:37 2.4Gb   100.0% Consistency (2/2)\n",
      "04:37 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb  Align cluster 9 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 10 / 348 (13 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% Consistency (1/2)\n",
      "04:38 2.4Gb   100.0% Consistency (2/2)\n",
      "04:38 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb  Align cluster 11 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 12 / 348 (7 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% Consistency (1/2)\n",
      "04:38 2.4Gb   100.0% Consistency (2/2)\n",
      "04:38 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb                  \n",
      "04:38 2.4Gb  Align cluster 13 / 348 (2 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% UPGMA5         \n",
      "04:38 2.4Gb  Align cluster 14 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 15 / 348 (2 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% UPGMA5         \n",
      "04:38 2.4Gb                \n",
      "04:38 2.4Gb  Align cluster 16 / 348 (11 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% Consistency (1/2)\n",
      "04:38 2.4Gb   100.0% Consistency (2/2)\n",
      "04:38 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb  Align cluster 17 / 348 (1 seq)\n",
      "04:38 2.4Gb  Align cluster 18 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 19 / 348 (3 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% Consistency (1/2)\n",
      "04:38 2.4Gb   100.0% Consistency (2/2)\n",
      "04:38 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb  Align cluster 20 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 21 / 348 (14 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:38 2.4Gb   100.0% Consistency (1/2)\n",
      "04:38 2.4Gb   100.0% Consistency (2/2)\n",
      "04:38 2.4Gb   100.0% UPGMA5           \n",
      "04:38 2.4Gb   100.0% Refining\n",
      "04:38 2.4Gb  Align cluster 22 / 348 (1 seq)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb  Align cluster 23 / 348 (51 seqs)\n",
      "04:38 2.4Gb  \n",
      "04:38 2.4Gb   100.0% Calc posteriors\n",
      "04:39 2.4Gb   100.0% Consistency (1/2)\n",
      "04:39 2.4Gb   100.0% Consistency (2/2)\n",
      "04:39 2.4Gb   100.0% UPGMA5           \n",
      "04:39 2.4Gb   100.0% Refining\n",
      "04:39 2.4Gb  Align cluster 24 / 348 (1 seq)\n",
      "04:39 2.4Gb  \n",
      "04:39 2.4Gb  Align cluster 25 / 348 (183 seqs)\n",
      "04:39 2.4Gb  \n",
      "04:40 2.4Gb   100.0% Calc posteriors\n",
      "04:42 2.4Gb   100.0% Consistency (1/2)\n",
      "04:44 2.4Gb   100.0% Consistency (2/2)\n",
      "04:44 2.4Gb   100.0% UPGMA5           \n",
      "04:47 2.4Gb   100.0% Refining\n",
      "04:47 2.4Gb                  \n",
      "04:47 2.4Gb  Align cluster 26 / 348 (15 seqs)\n",
      "04:47 2.4Gb  \n",
      "04:47 2.4Gb   100.0% Calc posteriors\n",
      "04:47 2.4Gb   100.0% Consistency (1/2)\n",
      "04:47 2.4Gb   100.0% Consistency (2/2)\n",
      "04:47 2.4Gb   100.0% UPGMA5           \n",
      "04:47 2.4Gb   100.0% Refining\n",
      "04:47 2.4Gb  Align cluster 27 / 348 (1 seq)\n",
      "04:47 2.4Gb  Align cluster 28 / 348 (1 seq)\n",
      "04:47 2.4Gb  Align cluster 29 / 348 (1 seq)\n",
      "04:47 2.4Gb  Align cluster 30 / 348 (1 seq)\n",
      "04:47 2.4Gb  \n",
      "04:47 2.4Gb  Align cluster 31 / 348 (8 seqs)\n",
      "04:47 2.4Gb  \n",
      "04:47 2.4Gb   100.0% Calc posteriors\n",
      "04:47 2.4Gb   100.0% Consistency (1/2)\n",
      "04:47 2.4Gb   100.0% Consistency (2/2)\n",
      "04:47 2.4Gb   100.0% UPGMA5           \n",
      "04:47 2.4Gb   100.0% Refining\n",
      "04:47 2.4Gb                  \n",
      "04:47 2.4Gb  Align cluster 32 / 348 (120 seqs)\n",
      "04:47 2.4Gb  \n",
      "04:48 2.4Gb   100.0% Calc posteriors\n",
      "04:48 2.4Gb   100.0% Consistency (1/2)\n",
      "04:49 2.4Gb   100.0% Consistency (2/2)\n",
      "04:49 2.4Gb   100.0% UPGMA5           \n",
      "04:50 2.4Gb   100.0% Refining\n",
      "04:50 2.4Gb  Align cluster 33 / 348 (1 seq)\n",
      "04:50 2.4Gb  \n",
      "04:50 2.4Gb  Align cluster 34 / 348 (2 seqs)\n",
      "04:50 2.4Gb  \n",
      "04:50 2.4Gb   100.0% Calc posteriors\n",
      "04:50 2.4Gb   100.0% UPGMA5         \n",
      "04:50 2.4Gb  Align cluster 35 / 348 (1 seq)\n",
      "04:50 2.4Gb  \n",
      "04:50 2.4Gb  Align cluster 36 / 348 (2 seqs)\n",
      "04:50 2.4Gb  \n",
      "04:50 2.4Gb   100.0% Calc posteriors\n",
      "04:50 2.4Gb   100.0% UPGMA5         \n",
      "04:50 2.4Gb                \n",
      "04:50 2.4Gb  Align cluster 37 / 348 (3 seqs)\n",
      "04:50 2.4Gb  \n",
      "04:50 2.4Gb   100.0% Calc posteriors\n",
      "04:50 2.4Gb   100.0% Consistency (1/2)\n",
      "04:50 2.4Gb   100.0% Consistency (2/2)\n",
      "04:50 2.4Gb   100.0% UPGMA5           \n",
      "04:50 2.4Gb   100.0% Refining\n",
      "04:50 2.4Gb                  \n",
      "04:50 2.4Gb  Align cluster 38 / 348 (52 seqs)\n",
      "04:50 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% Consistency (1/2)\n",
      "04:51 2.4Gb   100.0% Consistency (2/2)\n",
      "04:51 2.4Gb   100.0% UPGMA5           \n",
      "04:51 2.4Gb   100.0% Refining\n",
      "04:51 2.4Gb  Align cluster 39 / 348 (1 seq)\n",
      "04:51 2.4Gb  Align cluster 40 / 348 (1 seq)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb  Align cluster 41 / 348 (3 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% Consistency (1/2)\n",
      "04:51 2.4Gb   100.0% Consistency (2/2)\n",
      "04:51 2.4Gb   100.0% UPGMA5           \n",
      "04:51 2.4Gb   100.0% Refining\n",
      "04:51 2.4Gb                  \n",
      "04:51 2.4Gb  Align cluster 42 / 348 (6 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% Consistency (1/2)\n",
      "04:51 2.4Gb   100.0% Consistency (2/2)\n",
      "04:51 2.4Gb   100.0% UPGMA5           \n",
      "04:51 2.4Gb   100.0% Refining\n",
      "04:51 2.4Gb                  \n",
      "04:51 2.4Gb  Align cluster 43 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb  Align cluster 44 / 348 (1 seq)\n",
      "04:51 2.4Gb  Align cluster 45 / 348 (1 seq)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb  Align cluster 46 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb                \n",
      "04:51 2.4Gb  Align cluster 47 / 348 (12 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% Consistency (1/2)\n",
      "04:51 2.4Gb   100.0% Consistency (2/2)\n",
      "04:51 2.4Gb   100.0% UPGMA5           \n",
      "04:51 2.4Gb   100.0% Refining\n",
      "04:51 2.4Gb  Align cluster 48 / 348 (1 seq)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb  Align cluster 49 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb  Align cluster 50 / 348 (1 seq)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb  Align cluster 51 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb                \n",
      "04:51 2.4Gb  Align cluster 52 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb  Align cluster 53 / 348 (1 seq)\n",
      "04:51 2.4Gb  Align cluster 54 / 348 (1 seq)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb  Align cluster 55 / 348 (2 seqs)\n",
      "04:51 2.4Gb  \n",
      "04:51 2.4Gb   100.0% Calc posteriors\n",
      "04:51 2.4Gb   100.0% UPGMA5         \n",
      "04:51 2.4Gb                \n",
      "04:51 2.4Gb  Align cluster 56 / 348 (477 seqs)\n",
      "04:51 2.4Gb  \n",
      "05:10 2.4Gb   100.0% Calc posteriors  \n",
      "06:08 2.4Gb   100.0% Consistency (1/2)  \n",
      "06:46 2.4Gb   100.0% Consistency (2/2)  \n",
      "06:46 2.4Gb   100.0% UPGMA5           \n",
      "07:10 2.4Gb   100.0% Refining\n",
      "07:10 2.4Gb  Align cluster 57 / 348 (1 seq)\n",
      "07:10 2.4Gb  \n",
      "07:10 2.4Gb  Align cluster 58 / 348 (3 seqs)\n",
      "07:10 2.4Gb  \n",
      "07:10 2.4Gb   100.0% Calc posteriors\n",
      "07:10 2.4Gb   100.0% Consistency (1/2)\n",
      "07:10 2.4Gb   100.0% Consistency (2/2)\n",
      "07:10 2.4Gb   100.0% UPGMA5           \n",
      "07:10 2.4Gb   100.0% Refining\n",
      "07:10 2.4Gb  Align cluster 59 / 348 (1 seq)\n",
      "07:10 2.4Gb  \n",
      "07:10 2.4Gb  Align cluster 60 / 348 (5 seqs)\n",
      "07:10 2.4Gb  \n",
      "07:10 2.4Gb   100.0% Calc posteriors\n",
      "07:11 2.4Gb   100.0% Consistency (1/2)\n",
      "07:11 2.4Gb   100.0% Consistency (2/2)\n",
      "07:11 2.4Gb   100.0% UPGMA5           \n",
      "07:11 2.4Gb   100.0% Refining\n",
      "07:11 2.4Gb  Align cluster 61 / 348 (1 seq)\n",
      "07:11 2.4Gb  \n",
      "07:11 2.4Gb  Align cluster 62 / 348 (2 seqs)\n",
      "07:11 2.4Gb  \n",
      "07:11 2.4Gb   100.0% Calc posteriors\n",
      "07:11 2.4Gb   100.0% UPGMA5         \n",
      "07:11 2.4Gb  Align cluster 63 / 348 (1 seq)\n",
      "07:11 2.4Gb  Align cluster 64 / 348 (1 seq)\n",
      "07:11 2.4Gb  Align cluster 65 / 348 (1 seq)\n",
      "07:11 2.4Gb  Align cluster 66 / 348 (1 seq)\n",
      "07:11 2.4Gb  Align cluster 67 / 348 (1 seq)\n",
      "07:11 2.4Gb  Align cluster 68 / 348 (1 seq)\n",
      "07:11 2.4Gb  \n",
      "07:11 2.4Gb  Align cluster 69 / 348 (57 seqs)\n",
      "07:11 2.4Gb  \n",
      "07:11 2.4Gb   100.0% Calc posteriors\n",
      "07:11 2.4Gb   100.0% Consistency (1/2)\n",
      "07:11 2.4Gb   100.0% Consistency (2/2)\n",
      "07:11 2.4Gb   100.0% UPGMA5           \n",
      "07:11 2.4Gb   100.0% Refining\n",
      "07:11 2.4Gb  Align cluster 70 / 348 (1 seq)\n",
      "07:11 2.4Gb  \n",
      "07:11 2.4Gb  Align cluster 71 / 348 (181 seqs)\n",
      "07:11 2.4Gb  \n",
      "07:13 2.4Gb   100.0% Calc posteriors \n",
      "07:15 2.4Gb   100.0% Consistency (1/2) \n",
      "07:17 2.4Gb   100.0% Consistency (2/2) \n",
      "07:17 2.4Gb   100.0% UPGMA5           \n",
      "07:19 2.4Gb   100.0% Refining\n",
      "07:19 2.4Gb                  \n",
      "07:19 2.4Gb  Align cluster 72 / 348 (197 seqs)\n",
      "07:19 2.4Gb  \n",
      "07:21 2.4Gb   100.0% Calc posteriors \n",
      "07:24 2.4Gb   100.0% Consistency (1/2) \n",
      "07:27 2.4Gb   100.0% Consistency (2/2) \n",
      "07:27 2.4Gb   100.0% UPGMA5           \n",
      "07:30 2.4Gb   100.0% Refining\n",
      "07:30 2.4Gb  Align cluster 73 / 348 (1 seq)\n",
      "07:30 2.4Gb  Align cluster 74 / 348 (1 seq)\n",
      "07:30 2.4Gb  \n",
      "07:30 2.4Gb  Align cluster 75 / 348 (64 seqs)\n",
      "07:30 2.4Gb  \n",
      "07:30 2.4Gb   100.0% Calc posteriors\n",
      "07:30 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb                  \n",
      "07:31 2.4Gb  Align cluster 76 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 77 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 78 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 79 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 80 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 81 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 82 / 348 (21 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb                  \n",
      "07:31 2.4Gb  Align cluster 83 / 348 (4 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb                  \n",
      "07:31 2.4Gb  Align cluster 84 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 85 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 86 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 87 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 88 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 89 / 348 (5 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb  Align cluster 90 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 91 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 92 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 93 / 348 (3 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb  Align cluster 94 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 95 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 96 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 97 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 98 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 99 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 100 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb                \n",
      "07:31 2.4Gb  Align cluster 101 / 348 (6 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb                  \n",
      "07:31 2.4Gb  Align cluster 102 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 103 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 104 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 105 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 106 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 107 / 348 (13 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% Consistency (1/2)\n",
      "07:31 2.4Gb   100.0% Consistency (2/2)\n",
      "07:31 2.4Gb   100.0% UPGMA5           \n",
      "07:31 2.4Gb   100.0% Refining\n",
      "07:31 2.4Gb  Align cluster 108 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 109 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 110 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 111 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb                \n",
      "07:31 2.4Gb  Align cluster 112 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 113 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 114 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 115 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 116 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 117 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 118 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 119 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 120 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 121 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:31 2.4Gb   100.0% UPGMA5         \n",
      "07:31 2.4Gb  Align cluster 122 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 123 / 348 (1 seq)\n",
      "07:31 2.4Gb  Align cluster 124 / 348 (1 seq)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb  Align cluster 125 / 348 (2 seqs)\n",
      "07:31 2.4Gb  \n",
      "07:31 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% UPGMA5         \n",
      "07:32 2.4Gb                \n",
      "07:32 2.4Gb  Align cluster 126 / 348 (21 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 127 / 348 (17 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 128 / 348 (33 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb  Align cluster 129 / 348 (1 seq)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb  Align cluster 130 / 348 (12 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 131 / 348 (20 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 132 / 348 (21 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 133 / 348 (26 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:32 2.4Gb   100.0% UPGMA5           \n",
      "07:32 2.4Gb   100.0% Refining\n",
      "07:32 2.4Gb                  \n",
      "07:32 2.4Gb  Align cluster 134 / 348 (25 seqs)\n",
      "07:32 2.4Gb  \n",
      "07:32 2.4Gb   100.0% Calc posteriors\n",
      "07:32 2.4Gb   100.0% Consistency (1/2)\n",
      "07:32 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 135 / 348 (13 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 136 / 348 (18 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb  Align cluster 137 / 348 (1 seq)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb  Align cluster 138 / 348 (2 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% UPGMA5         \n",
      "07:33 2.4Gb                \n",
      "07:33 2.4Gb  Align cluster 139 / 348 (11 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 140 / 348 (7 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 141 / 348 (19 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 142 / 348 (11 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb  Align cluster 143 / 348 (1 seq)\n",
      "07:33 2.4Gb  Align cluster 144 / 348 (1 seq)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb  Align cluster 145 / 348 (14 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 146 / 348 (13 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 147 / 348 (23 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 148 / 348 (9 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 149 / 348 (9 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 150 / 348 (8 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 151 / 348 (17 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb  Align cluster 152 / 348 (1 seq)\n",
      "07:33 2.4Gb  Align cluster 153 / 348 (1 seq)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb  Align cluster 154 / 348 (8 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:33 2.4Gb   100.0% Refining\n",
      "07:33 2.4Gb                  \n",
      "07:33 2.4Gb  Align cluster 155 / 348 (30 seqs)\n",
      "07:33 2.4Gb  \n",
      "07:33 2.4Gb   100.0% Calc posteriors\n",
      "07:33 2.4Gb   100.0% Consistency (1/2)\n",
      "07:33 2.4Gb   100.0% Consistency (2/2)\n",
      "07:33 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 156 / 348 (5 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb  Align cluster 157 / 348 (1 seq)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb  Align cluster 158 / 348 (15 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 159 / 348 (6 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 160 / 348 (15 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb  Align cluster 161 / 348 (1 seq)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb  Align cluster 162 / 348 (12 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 163 / 348 (8 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 164 / 348 (2 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% UPGMA5         \n",
      "07:34 2.4Gb                \n",
      "07:34 2.4Gb  Align cluster 165 / 348 (5 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 166 / 348 (13 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 167 / 348 (3 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 168 / 348 (3 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 169 / 348 (11 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 170 / 348 (9 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 171 / 348 (3 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:34 2.4Gb   100.0% Refining\n",
      "07:34 2.4Gb                  \n",
      "07:34 2.4Gb  Align cluster 172 / 348 (2 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% UPGMA5         \n",
      "07:34 2.4Gb                \n",
      "07:34 2.4Gb  Align cluster 173 / 348 (16 seqs)\n",
      "07:34 2.4Gb  \n",
      "07:34 2.4Gb   100.0% Calc posteriors\n",
      "07:34 2.4Gb   100.0% Consistency (1/2)\n",
      "07:34 2.4Gb   100.0% Consistency (2/2)\n",
      "07:34 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 174 / 348 (11 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb  Align cluster 175 / 348 (1 seq)\n",
      "07:35 2.4Gb  Align cluster 176 / 348 (1 seq)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb  Align cluster 177 / 348 (12 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 178 / 348 (3 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 179 / 348 (3 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 180 / 348 (8 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 181 / 348 (7 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 182 / 348 (10 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 183 / 348 (18 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 184 / 348 (9 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb  Align cluster 185 / 348 (1 seq)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb  Align cluster 186 / 348 (7 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 187 / 348 (2 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% UPGMA5         \n",
      "07:35 2.4Gb  Align cluster 188 / 348 (1 seq)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb  Align cluster 189 / 348 (3 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 190 / 348 (17 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb  Align cluster 191 / 348 (1 seq)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb  Align cluster 192 / 348 (5 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 193 / 348 (5 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 194 / 348 (2 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% UPGMA5         \n",
      "07:35 2.4Gb                \n",
      "07:35 2.4Gb  Align cluster 195 / 348 (2 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% UPGMA5         \n",
      "07:35 2.4Gb                \n",
      "07:35 2.4Gb  Align cluster 196 / 348 (10 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:35 2.4Gb   100.0% Consistency (1/2)\n",
      "07:35 2.4Gb   100.0% Consistency (2/2)\n",
      "07:35 2.4Gb   100.0% UPGMA5           \n",
      "07:35 2.4Gb   100.0% Refining\n",
      "07:35 2.4Gb                  \n",
      "07:35 2.4Gb  Align cluster 197 / 348 (2 seqs)\n",
      "07:35 2.4Gb  \n",
      "07:35 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb  Align cluster 198 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 199 / 348 (6 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb  Align cluster 200 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 201 / 348 (6 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 202 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb  Align cluster 203 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 204 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 205 / 348 (9 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb  Align cluster 206 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 207 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 208 / 348 (3 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 209 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb                \n",
      "07:36 2.4Gb  Align cluster 210 / 348 (3 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb  Align cluster 211 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 212 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 213 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 214 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 215 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb                \n",
      "07:36 2.4Gb  Align cluster 216 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb  Align cluster 217 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 218 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb  Align cluster 219 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 220 / 348 (2 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% UPGMA5         \n",
      "07:36 2.4Gb  Align cluster 221 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 222 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 223 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 224 / 348 (1 seq)\n",
      "07:36 2.4Gb  Align cluster 225 / 348 (1 seq)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb  Align cluster 226 / 348 (15 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 227 / 348 (34 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 228 / 348 (5 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 229 / 348 (30 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:36 2.4Gb   100.0% Consistency (2/2)\n",
      "07:36 2.4Gb   100.0% UPGMA5           \n",
      "07:36 2.4Gb   100.0% Refining\n",
      "07:36 2.4Gb                  \n",
      "07:36 2.4Gb  Align cluster 230 / 348 (22 seqs)\n",
      "07:36 2.4Gb  \n",
      "07:36 2.4Gb   100.0% Calc posteriors\n",
      "07:36 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 231 / 348 (24 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 232 / 348 (2 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% UPGMA5         \n",
      "07:37 2.4Gb                \n",
      "07:37 2.4Gb  Align cluster 233 / 348 (9 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 234 / 348 (12 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 235 / 348 (18 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 236 / 348 (7 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 237 / 348 (7 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 238 / 348 (6 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb  Align cluster 239 / 348 (1 seq)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb  Align cluster 240 / 348 (5 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 241 / 348 (8 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:37 2.4Gb   100.0% Consistency (2/2)\n",
      "07:37 2.4Gb   100.0% UPGMA5           \n",
      "07:37 2.4Gb   100.0% Refining\n",
      "07:37 2.4Gb                  \n",
      "07:37 2.4Gb  Align cluster 242 / 348 (26 seqs)\n",
      "07:37 2.4Gb  \n",
      "07:37 2.4Gb   100.0% Calc posteriors\n",
      "07:37 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 243 / 348 (11 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 244 / 348 (6 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 245 / 348 (10 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)cy (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 246 / 348 (6 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb  Align cluster 247 / 348 (1 seq)\n",
      "07:38 2.4Gb  Align cluster 248 / 348 (1 seq)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb  Align cluster 249 / 348 (6 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 250 / 348 (18 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb  Align cluster 251 / 348 (1 seq)\n",
      "07:38 2.4Gb  Align cluster 252 / 348 (1 seq)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb  Align cluster 253 / 348 (2 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% UPGMA5         \n",
      "07:38 2.4Gb                \n",
      "07:38 2.4Gb  Align cluster 254 / 348 (6 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 255 / 348 (4 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 256 / 348 (3 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb  Align cluster 257 / 348 (1 seq)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb  Align cluster 258 / 348 (8 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb  Align cluster 259 / 348 (1 seq)\n",
      "07:38 2.4Gb  Align cluster 260 / 348 (1 seq)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb  Align cluster 261 / 348 (5 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 262 / 348 (6 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 263 / 348 (7 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 264 / 348 (3 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:38 2.4Gb   100.0% Calc posteriors\n",
      "07:38 2.4Gb   100.0% Consistency (1/2)\n",
      "07:38 2.4Gb   100.0% Consistency (2/2)\n",
      "07:38 2.4Gb   100.0% UPGMA5           \n",
      "07:38 2.4Gb   100.0% Refining\n",
      "07:38 2.4Gb                  \n",
      "07:38 2.4Gb  Align cluster 265 / 348 (7 seqs)\n",
      "07:38 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 266 / 348 (24 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 267 / 348 (15 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 268 / 348 (3 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 269 / 348 (4 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 270 / 348 (17 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 271 / 348 (6 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 272 / 348 (5 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 273 / 348 (9 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 274 / 348 (2 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% UPGMA5         \n",
      "07:39 2.4Gb                \n",
      "07:39 2.4Gb  Align cluster 275 / 348 (17 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb  Align cluster 276 / 348 (1 seq)\n",
      "07:39 2.4Gb  Align cluster 277 / 348 (1 seq)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb  Align cluster 278 / 348 (5 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 279 / 348 (2 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% UPGMA5         \n",
      "07:39 2.4Gb  Align cluster 280 / 348 (1 seq)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb  Align cluster 281 / 348 (5 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:39 2.4Gb   100.0% Consistency (2/2)\n",
      "07:39 2.4Gb   100.0% UPGMA5           \n",
      "07:39 2.4Gb   100.0% Refining\n",
      "07:39 2.4Gb                  \n",
      "07:39 2.4Gb  Align cluster 282 / 348 (3 seqs)\n",
      "07:39 2.4Gb  \n",
      "07:39 2.4Gb   100.0% Calc posteriors\n",
      "07:39 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 283 / 348 (5 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 284 / 348 (17 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb  Align cluster 285 / 348 (1 seq)\n",
      "07:40 2.4Gb  Align cluster 286 / 348 (1 seq)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb  Align cluster 287 / 348 (4 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 288 / 348 (4 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 289 / 348 (5 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 290 / 348 (5 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 291 / 348 (6 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 292 / 348 (7 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 293 / 348 (2 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% UPGMA5         \n",
      "07:40 2.4Gb                \n",
      "07:40 2.4Gb  Align cluster 294 / 348 (7 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb     1.0% Refining7:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 295 / 348 (10 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 296 / 348 (6 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb  Align cluster 297 / 348 (1 seq)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb  Align cluster 298 / 348 (2 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% UPGMA5         \n",
      "07:40 2.4Gb                \n",
      "07:40 2.4Gb  Align cluster 299 / 348 (7 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 300 / 348 (10 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb  Align cluster 301 / 348 (1 seq)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb  Align cluster 302 / 348 (5 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb  Align cluster 303 / 348 (1 seq)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb  Align cluster 304 / 348 (7 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:40 2.4Gb   100.0% UPGMA5           \n",
      "07:40 2.4Gb   100.0% Refining\n",
      "07:40 2.4Gb                  \n",
      "07:40 2.4Gb  Align cluster 305 / 348 (4 seqs)\n",
      "07:40 2.4Gb  \n",
      "07:40 2.4Gb   100.0% Calc posteriors\n",
      "07:40 2.4Gb   100.0% Consistency (1/2)\n",
      "07:40 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 306 / 348 (2 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% UPGMA5         \n",
      "07:41 2.4Gb                \n",
      "07:41 2.4Gb  Align cluster 307 / 348 (6 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 308 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 309 / 348 (2 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% UPGMA5         \n",
      "07:41 2.4Gb  Align cluster 310 / 348 (1 seq)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb  Align cluster 311 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 312 / 348 (4 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 313 / 348 (6 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 314 / 348 (2 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% UPGMA5         \n",
      "07:41 2.4Gb  Align cluster 315 / 348 (1 seq)\n",
      "07:41 2.4Gb  Align cluster 316 / 348 (1 seq)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb  Align cluster 317 / 348 (8 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 318 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 319 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 320 / 348 (6 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 321 / 348 (4 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 322 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 323 / 348 (2 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% UPGMA5         \n",
      "07:41 2.4Gb                \n",
      "07:41 2.4Gb  Align cluster 324 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb  Align cluster 325 / 348 (1 seq)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb  Align cluster 326 / 348 (3 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:41 2.4Gb   100.0% Consistency (1/2)\n",
      "07:41 2.4Gb   100.0% Consistency (2/2)\n",
      "07:41 2.4Gb   100.0% UPGMA5           \n",
      "07:41 2.4Gb   100.0% Refining\n",
      "07:41 2.4Gb                  \n",
      "07:41 2.4Gb  Align cluster 327 / 348 (2 seqs)\n",
      "07:41 2.4Gb  \n",
      "07:41 2.4Gb   100.0% Calc posteriors\n",
      "07:42 2.4Gb   100.0% UPGMA5         \n",
      "07:42 2.4Gb  Align cluster 328 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 329 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 330 / 348 (1 seq)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb  Align cluster 331 / 348 (2 seqs)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb   100.0% Calc posteriors\n",
      "07:42 2.4Gb   100.0% UPGMA5         \n",
      "07:42 2.4Gb                \n",
      "07:42 2.4Gb  Align cluster 332 / 348 (2 seqs)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb   100.0% Calc posteriors\n",
      "07:42 2.4Gb   100.0% UPGMA5         \n",
      "07:42 2.4Gb  Align cluster 333 / 348 (1 seq)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb  Align cluster 334 / 348 (2 seqs)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb   100.0% Calc posteriors\n",
      "07:42 2.4Gb   100.0% UPGMA5         \n",
      "07:42 2.4Gb                \n",
      "07:42 2.4Gb  Align cluster 335 / 348 (2 seqs)\n",
      "07:42 2.4Gb  \n",
      "07:42 2.4Gb   100.0% Calc posteriors\n",
      "07:42 2.4Gb   100.0% UPGMA5         \n",
      "07:42 2.4Gb  Align cluster 336 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 337 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 338 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 339 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 340 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 341 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 342 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 343 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 344 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 345 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 346 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 347 / 348 (1 seq)\n",
      "07:42 2.4Gb  Align cluster 348 / 348 (1 seq)\n",
      "07:42 2.4Gb   100.0% Consensus sequences\n",
      "07:48 2.4Gb   100.0% 348 consensus seqs, mean EE 0.4 \n",
      "07:48 2.4Gb   100.0% UPGMA5                         \n",
      "07:48 2.4Gb   100.0% Join 1 / 347 [1 x 1, 1 pairs]\n",
      "07:48 2.4Gb   100.0% Join 2 / 347 [1 x 1, 1 pairs]\n",
      "07:48 2.4Gb   100.0% Join 3 / 347 [1 x 2, 2 pairs]\n",
      "07:48 2.4Gb   100.0% Join 4 / 347 [1 x 2, 2 pairs]\n",
      "07:48 2.4Gb   100.0% Join 5 / 347 [1 x 7, 7 pairs]\n",
      "07:48 2.4Gb   100.0% Join 6 / 347 [3 x 8, 24 pairs]\n",
      "07:48 2.4Gb   100.0% Join 7 / 347 [6 x 6, 36 pairs]\n",
      "07:48 2.4Gb   100.0% Join 8 / 347 [1 x 12, 12 pairs]\n",
      "07:48 2.4Gb   100.0% Join 9 / 347 [6 x 13, 78 pairs]\n",
      "07:48 2.4Gb   100.0% Join 10 / 347 [17 x 17, 289 pairs]\n",
      "07:48 2.4Gb   100.0% Join 11 / 347 [2 x 34, 68 pairs]  \n",
      "07:48 2.4Gb   100.0% Join 12 / 347 [18 x 36, 648 pairs]\n",
      "07:48 2.4Gb   100.0% Join 13 / 347 [3 x 54, 162 pairs] \n",
      "07:48 2.4Gb   100.0% Join 14 / 347 [1 x 57, 57 pairs] \n",
      "07:48 2.4Gb   100.0% Join 15 / 347 [2 x 58, 116 pairs]\n",
      "07:48 2.4Gb   100.0% Join 16 / 347 [1 x 60, 60 pairs] \n",
      "07:48 2.4Gb   100.0% Join 17 / 347 [19 x 61, 1159 pairs]\n",
      "07:48 2.4Gb   100.0% Join 18 / 347 [2 x 80, 160 pairs]  \n",
      "07:48 2.4Gb   100.0% Join 19 / 347 [2 x 82, 164 pairs]\n",
      "07:48 2.4Gb   100.0% Join 20 / 347 [1 x 14, 14 pairs] \n",
      "07:48 2.4Gb   100.0% Join 21 / 347 [15 x 21, 315 pairs]\n",
      "07:48 2.4Gb   100.0% Join 22 / 347 [1 x 36, 36 pairs]  \n",
      "07:49 2.4Gb   100.0% Join 23 / 347 [3 x 5, 15 pairs] \n",
      "07:49 2.4Gb   100.0% Join 24 / 347 [3 x 8, 24 pairs]\n",
      "07:49 2.4Gb   100.0% Join 25 / 347 [1 x 11, 11 pairs]\n",
      "07:49 2.4Gb   100.0% Join 26 / 347 [12 x 23, 276 pairs]\n",
      "07:49 2.4Gb   100.0% Join 27 / 347 [6 x 12, 72 pairs]  \n",
      "07:49 2.4Gb   100.0% Join 28 / 347 [2 x 18, 36 pairs]\n",
      "07:49 2.4Gb   100.0% Join 29 / 347 [9 x 26, 234 pairs]\n",
      "07:49 2.4Gb   100.0% Join 30 / 347 [12 x 35, 420 pairs]\n",
      "07:49 2.4Gb   100.0% Join 31 / 347 [1 x 47, 47 pairs]  \n",
      "07:49 2.4Gb   100.0% Join 32 / 347 [10 x 17, 170 pairs]\n",
      "07:49 2.4Gb   100.0% Join 33 / 347 [13 x 13, 169 pairs]\n",
      "07:49 2.4Gb   100.0% Join 34 / 347 [26 x 27, 702 pairs]\n",
      "07:49 2.4Gb   100.0% Join 35 / 347 [3 x 53, 159 pairs] \n",
      "07:49 2.4Gb   100.0% Join 36 / 347 [3 x 10, 30 pairs] \n",
      "07:49 2.4Gb   100.0% Join 37 / 347 [7 x 34, 238 pairs]\n",
      "07:49 2.4Gb   100.0% Join 38 / 347 [6 x 41, 246 pairs]\n",
      "07:49 2.4Gb   100.0% Join 39 / 347 [4 x 21, 84 pairs] \n",
      "07:50 2.4Gb   100.0% Join 40 / 347 [25 x 47, 1175 pairs]\n",
      "07:50 2.4Gb   100.0% Join 41 / 347 [8 x 72, 576 pairs]  \n",
      "07:50 2.4Gb   100.0% Join 42 / 347 [5 x 80, 400 pairs]\n",
      "07:50 2.4Gb   100.0% Join 43 / 347 [13 x 85, 1105 pairs]\n",
      "07:50 2.4Gb   100.0% Join 44 / 347 [6 x 12, 72 pairs]   \n",
      "07:50 2.4Gb   100.0% Join 45 / 347 [1 x 61, 61 pairs]\n",
      "07:50 2.4Gb   100.0% Join 46 / 347 [17 x 17, 289 pairs]\n",
      "07:50 2.4Gb   100.0% Join 47 / 347 [34 x 62, 2108 pairs]\n",
      "07:51 2.4Gb   100.0% Join 48 / 347 [15 x 96, 1440 pairs]\n",
      "07:51 2.4Gb   100.0% Join 49 / 347 [83 x 111, 2000 pairs]\n",
      "07:51 2.4Gb   100.0% Join 50 / 347 [7 x 194, 1358 pairs] \n",
      "07:51 2.4Gb   100.0% Join 51 / 347 [9 x 201, 1809 pairs]\n",
      "07:52 2.4Gb   100.0% Join 52 / 347 [5 x 210, 1050 pairs]\n",
      "07:52 2.4Gb   100.0% Join 53 / 347 [3 x 215, 645 pairs] \n",
      "07:52 2.4Gb   100.0% Join 54 / 347 [18 x 218, 2000 pairs]\n",
      "07:52 2.4Gb   100.0% Join 55 / 347 [4 x 236, 944 pairs]  \n",
      "07:52 2.4Gb   100.0% Join 56 / 347 [98 x 240, 2000 pairs]\n",
      "07:52 2.4Gb   100.0% Join 57 / 347 [2 x 338, 676 pairs]  \n",
      "07:53 2.4Gb   100.0% Join 58 / 347 [56 x 340, 2000 pairs]\n",
      "07:53 2.4Gb   100.0% Join 59 / 347 [2 x 396, 792 pairs]  \n",
      "07:53 2.4Gb   100.0% Join 60 / 347 [1 x 398, 398 pairs]\n",
      "07:53 2.4Gb   100.0% Join 61 / 347 [2 x 399, 798 pairs]\n",
      "07:53 2.4Gb   100.0% Join 62 / 347 [183 x 401, 2000 pairs]\n",
      "07:54 2.4Gb   100.0% Join 63 / 347 [25 x 584, 2000 pairs] \n",
      "07:54 2.4Gb   100.0% Join 64 / 347 [9 x 609, 2000 pairs] \n",
      "07:54 2.4Gb   100.0% Join 65 / 347 [6 x 618, 2000 pairs]\n",
      "07:55 2.4Gb   100.0% Join 66 / 347 [48 x 624, 2000 pairs]\n",
      "07:55 2.4Gb   100.0% Join 67 / 347 [20 x 672, 2000 pairs]\n",
      "07:55 2.4Gb   100.0% Join 68 / 347 [35 x 692, 2000 pairs]\n",
      "07:55 2.4Gb   100.0% Join 69 / 347 [37 x 727, 2000 pairs]\n",
      "07:56 2.4Gb   100.0% Join 70 / 347 [84 x 764, 2000 pairs]\n",
      "07:56 2.4Gb   100.0% Join 71 / 347 [2 x 848, 1696 pairs] \n",
      "07:56 2.4Gb   100.0% Join 72 / 347 [1 x 850, 850 pairs] \n",
      "07:56 2.4Gb   100.0% Join 73 / 347 [2 x 851, 1702 pairs]\n",
      "07:56 2.4Gb   100.0% Join 74 / 347 [1 x 853, 853 pairs] \n",
      "07:57 2.4Gb   100.0% Join 75 / 347 [1 x 854, 854 pairs]\n",
      "07:57 2.4Gb   100.0% Join 76 / 347 [1 x 855, 855 pairs]\n",
      "07:57 2.4Gb   100.0% Join 77 / 347 [1 x 856, 856 pairs]\n",
      "07:57 2.4Gb   100.0% Join 78 / 347 [1 x 857, 857 pairs]\n",
      "07:57 2.4Gb   100.0% Join 79 / 347 [2 x 858, 1716 pairs]\n",
      "07:57 2.4Gb   100.0% Join 80 / 347 [1 x 860, 860 pairs] \n",
      "07:57 2.4Gb   100.0% Join 81 / 347 [1 x 861, 861 pairs]\n",
      "07:58 2.4Gb   100.0% Join 82 / 347 [1 x 862, 862 pairs]\n",
      "07:58 2.4Gb   100.0% Join 83 / 347 [1 x 863, 863 pairs]\n",
      "07:58 2.4Gb   100.0% Join 84 / 347 [1 x 864, 864 pairs]\n",
      "07:58 2.4Gb   100.0% Join 85 / 347 [1 x 865, 865 pairs]\n",
      "07:58 2.4Gb   100.0% Join 86 / 347 [1 x 866, 866 pairs]\n",
      "07:58 2.4Gb   100.0% Join 87 / 347 [1 x 867, 867 pairs]\n",
      "07:58 2.4Gb   100.0% Join 88 / 347 [1 x 868, 868 pairs]\n",
      "07:58 2.4Gb   100.0% Join 89 / 347 [1 x 869, 869 pairs]\n",
      "07:59 2.4Gb   100.0% Join 90 / 347 [1 x 870, 870 pairs]\n",
      "07:59 2.4Gb   100.0% Join 91 / 347 [1 x 871, 871 pairs]\n",
      "07:59 2.4Gb   100.0% Join 92 / 347 [1 x 872, 872 pairs]\n",
      "07:59 2.4Gb   100.0% Join 93 / 347 [1 x 873, 873 pairs]\n",
      "07:59 2.4Gb   100.0% Join 94 / 347 [1 x 874, 874 pairs]\n",
      "07:59 2.4Gb   100.0% Join 95 / 347 [11 x 875, 2000 pairs]\n",
      "07:59 2.4Gb   100.0% Join 96 / 347 [1 x 886, 886 pairs]  \n",
      "08:00 2.4Gb   100.0% Join 97 / 347 [1 x 887, 887 pairs]\n",
      "08:00 2.4Gb   100.0% Join 98 / 347 [1 x 888, 888 pairs]\n",
      "08:00 2.4Gb   100.0% Join 99 / 347 [1 x 889, 889 pairs]\n",
      "08:00 2.4Gb   100.0% Join 100 / 347 [2 x 890, 1780 pairs]\n",
      "08:00 2.4Gb   100.0% Join 101 / 347 [1 x 892, 892 pairs] \n",
      "08:00 2.4Gb   100.0% Join 102 / 347 [2 x 893, 1786 pairs]\n",
      "08:00 2.4Gb   100.0% Join 103 / 347 [1 x 895, 895 pairs] \n",
      "08:01 2.4Gb   100.0% Join 104 / 347 [1 x 896, 896 pairs]\n",
      "08:01 2.4Gb   100.0% Join 105 / 347 [1 x 897, 897 pairs]\n",
      "08:01 2.4Gb   100.0% Join 106 / 347 [1 x 898, 898 pairs]\n",
      "08:01 2.4Gb   100.0% Join 107 / 347 [3 x 899, 2697 pairs]\n",
      "08:01 2.4Gb   100.0% Join 108 / 347 [1 x 902, 902 pairs] \n",
      "08:01 2.4Gb   100.0% Join 109 / 347 [1 x 903, 903 pairs]\n",
      "08:02 2.4Gb   100.0% Join 110 / 347 [3 x 904, 2712 pairs]\n",
      "08:02 2.4Gb   100.0% Join 111 / 347 [1 x 907, 907 pairs] \n",
      "08:02 2.4Gb   100.0% Join 112 / 347 [2 x 908, 1816 pairs]\n",
      "08:02 2.4Gb   100.0% Join 113 / 347 [2 x 910, 1820 pairs]\n",
      "08:03 2.4Gb   100.0% Join 114 / 347 [1 x 912, 912 pairs] \n",
      "08:03 2.4Gb   100.0% Join 115 / 347 [1 x 913, 913 pairs]\n",
      "08:03 2.4Gb   100.0% Join 116 / 347 [1 x 3, 3 pairs]    \n",
      "08:03 2.4Gb   100.0% Join 117 / 347 [1 x 4, 4 pairs]\n",
      "08:03 2.4Gb   100.0% Join 118 / 347 [1 x 1, 1 pairs]\n",
      "08:03 2.4Gb   100.0% Join 119 / 347 [3 x 3, 9 pairs]\n",
      "08:03 2.4Gb   100.0% Join 120 / 347 [2 x 3, 6 pairs]\n",
      "08:03 2.4Gb   100.0% Join 121 / 347 [5 x 6, 30 pairs]\n",
      "08:03 2.4Gb   100.0% Join 122 / 347 [1 x 11, 11 pairs]\n",
      "08:03 2.4Gb   100.0% Join 123 / 347 [1 x 4, 4 pairs]  \n",
      "08:03 2.4Gb   100.0% Join 124 / 347 [1 x 2, 2 pairs]\n",
      "08:03 2.4Gb   100.0% Join 125 / 347 [3 x 28, 84 pairs]\n",
      "08:03 2.4Gb   100.0% Join 126 / 347 [5 x 31, 155 pairs]\n",
      "08:03 2.4Gb   100.0% Join 127 / 347 [1 x 36, 36 pairs] \n",
      "08:03 2.4Gb   100.0% Join 128 / 347 [1 x 3, 3 pairs]  \n",
      "08:03 2.4Gb   100.0% Join 129 / 347 [1 x 1, 1 pairs]\n",
      "08:03 2.4Gb   100.0% Join 130 / 347 [3 x 9, 27 pairs]\n",
      "08:03 2.4Gb   100.0% Join 131 / 347 [5 x 12, 60 pairs]\n",
      "08:03 2.4Gb   100.0% Join 132 / 347 [1 x 5, 5 pairs]  \n",
      "08:03 2.4Gb   100.0% Join 133 / 347 [1 x 6, 6 pairs]\n",
      "08:03 2.4Gb   100.0% Join 134 / 347 [1 x 7, 7 pairs]\n",
      "08:03 2.4Gb   100.0% Join 135 / 347 [1 x 8, 8 pairs]\n",
      "08:03 2.4Gb   100.0% Join 136 / 347 [4 x 4, 16 pairs]\n",
      "08:03 2.4Gb   100.0% Join 137 / 347 [15 x 18, 270 pairs]\n",
      "08:03 2.4Gb   100.0% Join 138 / 347 [24 x 33, 792 pairs]\n",
      "08:03 2.4Gb   100.0% Join 139 / 347 [7 x 57, 399 pairs] \n",
      "08:03 2.4Gb   100.0% Join 140 / 347 [8 x 13, 104 pairs]\n",
      "08:03 2.4Gb   100.0% Join 141 / 347 [21 x 64, 1344 pairs]\n",
      "08:04 2.4Gb   100.0% Join 142 / 347 [7 x 85, 595 pairs]  \n",
      "08:04 2.4Gb   100.0% Join 143 / 347 [64 x 92, 2000 pairs]\n",
      "08:04 2.4Gb   100.0% Join 144 / 347 [33 x 156, 2000 pairs]\n",
      "08:04 2.4Gb   100.0% Join 145 / 347 [1 x 189, 189 pairs]  \n",
      "08:04 2.4Gb   100.0% Join 146 / 347 [8 x 190, 1520 pairs]\n",
      "08:04 2.4Gb   100.0% Join 147 / 347 [7 x 198, 1386 pairs]\n",
      "08:04 2.4Gb   100.0% Join 148 / 347 [3 x 18, 54 pairs]   \n",
      "08:05 2.4Gb   100.0% Join 149 / 347 [21 x 207, 2000 pairs]\n",
      "08:05 2.4Gb   100.0% Join 150 / 347 [3 x 228, 684 pairs]  \n",
      "08:05 2.4Gb   100.0% Join 151 / 347 [52 x 231, 2000 pairs]\n",
      "08:05 2.4Gb   100.0% Join 152 / 347 [6 x 283, 1698 pairs] \n",
      "08:05 2.4Gb   100.0% Join 153 / 347 [4 x 5, 20 pairs]    \n",
      "08:05 2.4Gb   100.0% Join 154 / 347 [2 x 3, 6 pairs] \n",
      "08:05 2.4Gb   100.0% Join 155 / 347 [1 x 2, 2 pairs]\n",
      "08:05 2.4Gb   100.0% Join 156 / 347 [2 x 6, 12 pairs]\n",
      "08:05 2.4Gb   100.0% Join 157 / 347 [3 x 8, 24 pairs]\n",
      "08:05 2.4Gb   100.0% Join 158 / 347 [6 x 11, 66 pairs]\n",
      "08:05 2.4Gb   100.0% Join 159 / 347 [5 x 17, 85 pairs]\n",
      "08:05 2.4Gb   100.0% Join 160 / 347 [9 x 22, 198 pairs]\n",
      "08:05 2.4Gb   100.0% Join 161 / 347 [31 x 289, 2000 pairs]\n",
      "08:05 2.4Gb   100.0% Join 162 / 347 [5 x 11, 55 pairs]    \n",
      "08:05 2.4Gb   100.0% Join 163 / 347 [7 x 26, 182 pairs]\n",
      "08:05 2.4Gb   100.0% Join 164 / 347 [2 x 33, 66 pairs] \n",
      "08:05 2.4Gb   100.0% Join 165 / 347 [7 x 9, 63 pairs] \n",
      "08:05 2.4Gb   100.0% Join 166 / 347 [3 x 16, 48 pairs]\n",
      "08:06 2.4Gb   100.0% Join 167 / 347 [19 x 181, 2000 pairs]\n",
      "08:06 2.4Gb   100.0% Join 168 / 347 [12 x 200, 2400 pairs]\n",
      "08:06 2.4Gb   100.0% Join 169 / 347 [9 x 11, 99 pairs]    \n",
      "08:06 2.4Gb   100.0% Join 170 / 347 [8 x 20, 160 pairs]\n",
      "08:06 2.4Gb   100.0% Join 171 / 347 [3 x 6, 18 pairs]  \n",
      "08:06 2.4Gb   100.0% Join 172 / 347 [2 x 9, 18 pairs]\n",
      "08:06 2.4Gb   100.0% Join 173 / 347 [7 x 11, 77 pairs]\n",
      "08:06 2.4Gb   100.0% Join 174 / 347 [18 x 22, 396 pairs]\n",
      "08:06 2.4Gb   100.0% Join 175 / 347 [40 x 477, 2000 pairs]\n",
      "08:07 2.4Gb   100.0% Join 176 / 347 [120 x 517, 2000 pairs]\n",
      "08:07 2.4Gb   100.0% Join 177 / 347 [15 x 637, 2000 pairs] \n",
      "08:07 2.4Gb   100.0% Join 178 / 347 [4 x 652, 2608 pairs] \n",
      "08:07 2.4Gb   100.0% Join 179 / 347 [11 x 12, 132 pairs] \n",
      "08:07 2.4Gb   100.0% Join 180 / 347 [21 x 30, 630 pairs]\n",
      "08:07 2.4Gb   100.0% Join 181 / 347 [3 x 6, 18 pairs]   \n",
      "08:07 2.4Gb   100.0% Join 182 / 347 [10 x 10, 100 pairs]\n",
      "08:07 2.4Gb   100.0% Join 183 / 347 [7 x 20, 140 pairs] \n",
      "08:07 2.4Gb   100.0% Join 184 / 347 [9 x 27, 243 pairs]\n",
      "08:08 2.4Gb   100.0% Join 185 / 347 [36 x 51, 1836 pairs]\n",
      "08:08 2.4Gb   100.0% Join 186 / 347 [23 x 87, 2001 pairs]\n",
      "08:08 2.4Gb   100.0% Join 187 / 347 [20 x 110, 2200 pairs]\n",
      "08:08 2.4Gb   100.0% Join 188 / 347 [16 x 130, 2080 pairs]\n",
      "08:08 2.4Gb   100.0% Join 189 / 347 [5 x 10, 50 pairs]    \n",
      "08:08 2.4Gb   100.0% Join 190 / 347 [2 x 9, 18 pairs] \n",
      "08:08 2.4Gb   100.0% Join 191 / 347 [1 x 11, 11 pairs]\n",
      "08:08 2.4Gb   100.0% Join 192 / 347 [5 x 12, 60 pairs]\n",
      "08:08 2.4Gb   100.0% Join 193 / 347 [4 x 17, 68 pairs]\n",
      "08:08 2.4Gb   100.0% Join 194 / 347 [1 x 21, 21 pairs]\n",
      "08:08 2.4Gb   100.0% Join 195 / 347 [18 x 30, 540 pairs]\n",
      "08:08 2.4Gb   100.0% Join 196 / 347 [2 x 6, 12 pairs]   \n",
      "08:08 2.4Gb   100.0% Join 197 / 347 [3 x 8, 24 pairs]\n",
      "08:08 2.4Gb   100.0% Join 198 / 347 [5 x 5, 25 pairs]\n",
      "08:08 2.4Gb   10.91% Join 199 / 347 [10 x 11, 110 pairs]0.0% Join 199 / 347 [10 x 11, 110 pairs]\n",
      "08:09 2.4Gb   100.0% Join 200 / 347 [21 x 48, 1008 pairs]\n",
      "08:09 2.4Gb   100.0% Join 201 / 347 [22 x 69, 1518 pairs]\n",
      "08:09 2.4Gb   100.0% Join 202 / 347 [19 x 91, 1729 pairs]\n",
      "08:09 2.4Gb   100.0% Join 203 / 347 [15 x 110, 1650 pairs]\n",
      "08:10 2.4Gb   100.0% Join 204 / 347 [125 x 146, 2000 pairs]\n",
      "08:10 2.4Gb   100.0% Join 205 / 347 [2 x 271, 542 pairs]   \n",
      "08:10 2.4Gb   100.0% Join 206 / 347 [273 x 656, 2000 pairs]\n",
      "08:10 2.4Gb   100.0% Join 207 / 347 [24 x 929, 2000 pairs] \n",
      "08:10 2.4Gb   100.0% Join 208 / 347 [11 x 953, 2000 pairs]\n",
      "08:11 2.4Gb   100.0% Join 209 / 347 [28 x 964, 2000 pairs]\n",
      "08:11 2.4Gb   100.0% Join 210 / 347 [7 x 992, 2000 pairs] \n",
      "08:11 2.4Gb   100.0% Join 211 / 347 [14 x 999, 2000 pairs]\n",
      "08:12 2.4Gb   100.0% Join 212 / 347 [1 x 1013, 1013 pairs]\n",
      "08:12 2.4Gb   100.0% Join 213 / 347 [5 x 1014, 2000 pairs]\n",
      "08:12 2.4Gb   100.0% Join 214 / 347 [212 x 1019, 2000 pairs]\n",
      "08:12 2.4Gb   100.0% Join 215 / 347 [35 x 1231, 2000 pairs] \n",
      "08:13 2.4Gb   100.0% Join 216 / 347 [3 x 1266, 2000 pairs] \n",
      "08:13 2.4Gb   100.0% Join 217 / 347 [16 x 1269, 2000 pairs]\n",
      "08:13 2.4Gb   100.0% Join 218 / 347 [1 x 1285, 1285 pairs] \n",
      "08:13 2.4Gb   100.0% Join 219 / 347 [320 x 1286, 2000 pairs]\n",
      "08:14 2.4Gb   100.0% Join 220 / 347 [1 x 1606, 1606 pairs]  \n",
      "08:14 2.4Gb   100.0% Join 221 / 347 [2 x 1607, 2000 pairs]\n",
      "08:14 2.4Gb   100.0% Join 222 / 347 [205 x 1609, 2000 pairs]\n",
      "08:15 2.4Gb   100.0% Join 223 / 347 [1 x 1814, 1814 pairs]  \n",
      "08:15 2.4Gb   100.0% Join 224 / 347 [8 x 1815, 2000 pairs]\n",
      "08:15 2.4Gb   100.0% Join 225 / 347 [2 x 1823, 2000 pairs]\n",
      "08:15 2.4Gb   100.0% Join 226 / 347 [6 x 1825, 2000 pairs]\n",
      "08:16 2.4Gb   100.0% Join 227 / 347 [1 x 1831, 1831 pairs]\n",
      "08:16 2.4Gb   100.0% Join 228 / 347 [9 x 1832, 2000 pairs]\n",
      "08:16 2.4Gb   100.0% Join 229 / 347 [17 x 1841, 2000 pairs]\n",
      "08:16 2.4Gb   100.0% Join 230 / 347 [1 x 1858, 1858 pairs] \n",
      "08:17 2.4Gb   100.0% Join 231 / 347 [1 x 1859, 1859 pairs]\n",
      "08:17 2.4Gb   100.0% Join 232 / 347 [1 x 1860, 1860 pairs]\n",
      "08:17 2.4Gb   100.0% Join 233 / 347 [1 x 1861, 1861 pairs]\n",
      "08:17 2.4Gb   100.0% Join 234 / 347 [4 x 1862, 2000 pairs]\n",
      "08:18 2.4Gb   100.0% Join 235 / 347 [1 x 1866, 1866 pairs]\n",
      "08:18 2.4Gb   100.0% Join 236 / 347 [1 x 1867, 1867 pairs]\n",
      "08:18 2.4Gb   100.0% Join 237 / 347 [1 x 1868, 1868 pairs]\n",
      "08:18 2.4Gb   100.0% Join 238 / 347 [1 x 1869, 1869 pairs]\n",
      "08:19 2.4Gb   100.0% Join 239 / 347 [1 x 1870, 1870 pairs]\n",
      "08:19 2.4Gb   100.0% Join 240 / 347 [1 x 1871, 1871 pairs]\n",
      "08:19 2.4Gb   100.0% Join 241 / 347 [1 x 1872, 1872 pairs]\n",
      "08:20 2.4Gb   100.0% Join 242 / 347 [1 x 1873, 1873 pairs]\n",
      "08:20 2.4Gb   100.0% Join 243 / 347 [15 x 1874, 2000 pairs]\n",
      "08:20 2.4Gb   100.0% Join 244 / 347 [2 x 1889, 2000 pairs] \n",
      "08:20 2.4Gb   100.0% Join 245 / 347 [1 x 1891, 1891 pairs]\n",
      "08:21 2.4Gb   100.0% Join 246 / 347 [2 x 1892, 2000 pairs]\n",
      "08:21 2.4Gb   100.0% Join 247 / 347 [2 x 1894, 2000 pairs]\n",
      "08:21 2.4Gb   100.0% Join 248 / 347 [1 x 1896, 1896 pairs]\n",
      "08:22 2.4Gb   100.0% Join 249 / 347 [2 x 1897, 2000 pairs]\n",
      "08:22 2.4Gb   100.0% Join 250 / 347 [13 x 1899, 2000 pairs]\n",
      "08:22 2.4Gb   100.0% Join 251 / 347 [6 x 1912, 2000 pairs] \n",
      "08:22 2.4Gb   100.0% Join 252 / 347 [1 x 1918, 1918 pairs]\n",
      "08:23 2.4Gb   100.0% Join 253 / 347 [2 x 1919, 2000 pairs]\n",
      "08:23 2.4Gb   100.0% Join 254 / 347 [2 x 1921, 2000 pairs]\n",
      "08:23 2.4Gb   100.0% Join 255 / 347 [1 x 1923, 1923 pairs]\n",
      "08:24 2.4Gb   100.0% Join 256 / 347 [1 x 1924, 1924 pairs]\n",
      "08:24 2.4Gb   100.0% Join 257 / 347 [2 x 1925, 2000 pairs]\n",
      "08:24 2.4Gb   100.0% Join 258 / 347 [2 x 1927, 2000 pairs]\n",
      "08:25 2.4Gb   100.0% Join 259 / 347 [1 x 1929, 1929 pairs]\n",
      "08:25 2.4Gb   100.0% Join 260 / 347 [4 x 1930, 2000 pairs]\n",
      "08:25 2.4Gb   100.0% Join 261 / 347 [1 x 1934, 1934 pairs]\n",
      "08:26 2.4Gb   100.0% Join 262 / 347 [1 x 1935, 1935 pairs]\n",
      "08:26 2.4Gb   100.0% Join 263 / 347 [2 x 1936, 2000 pairs]\n",
      "08:26 2.4Gb   100.0% Join 264 / 347 [37 x 1938, 2000 pairs]\n",
      "08:26 2.4Gb   100.0% Join 265 / 347 [1 x 1975, 1975 pairs] \n",
      "08:26 2.4Gb   100.0% Join 266 / 347 [1 x 1, 1 pairs]      \n",
      "08:27 2.4Gb   100.0% Join 267 / 347 [2 x 5, 10 pairs]\n",
      "08:27 2.4Gb   100.0% Join 268 / 347 [1 x 1, 1 pairs] \n",
      "08:27 2.4Gb   100.0% Join 269 / 347 [1 x 2, 2 pairs]\n",
      "08:27 2.4Gb   100.0% Join 270 / 347 [3 x 7, 21 pairs]\n",
      "08:27 2.4Gb   100.0% Join 271 / 347 [1 x 5, 5 pairs] \n",
      "08:27 2.4Gb   100.0% Join 272 / 347 [6 x 8, 48 pairs]\n",
      "08:27 2.4Gb   100.0% Join 273 / 347 [1 x 1, 1 pairs] \n",
      "08:27 2.4Gb   100.0% Join 274 / 347 [1 x 2, 2 pairs]\n",
      "08:27 2.4Gb   100.0% Join 275 / 347 [1 x 3, 3 pairs]\n",
      "08:27 2.4Gb   100.0% Join 276 / 347 [3 x 51, 153 pairs]\n",
      "08:27 2.4Gb   100.0% Join 277 / 347 [13 x 54, 702 pairs]\n",
      "08:27 2.4Gb   100.0% Join 278 / 347 [2 x 2, 4 pairs]    \n",
      "08:27 2.4Gb   100.0% Join 279 / 347 [5 x 11, 55 pairs]\n",
      "08:27 2.4Gb   100.0% Join 280 / 347 [1 x 16, 16 pairs]\n",
      "08:27 2.4Gb   100.0% Join 281 / 347 [4 x 17, 68 pairs]\n",
      "08:27 2.4Gb   100.0% Join 282 / 347 [1 x 21, 21 pairs]\n",
      "08:27 2.4Gb   100.0% Join 283 / 347 [6 x 22, 132 pairs]\n",
      "08:27 2.4Gb   100.0% Join 284 / 347 [1 x 28, 28 pairs] \n",
      "08:27 2.4Gb   100.0% Join 285 / 347 [29 x 67, 1943 pairs]\n",
      "08:27 2.4Gb   100.0% Join 286 / 347 [1 x 96, 96 pairs]   \n",
      "08:27 2.4Gb   100.0% Join 287 / 347 [2 x 2, 4 pairs]  \n",
      "08:27 2.4Gb   100.0% Join 288 / 347 [4 x 8, 32 pairs]\n",
      "08:27 2.4Gb   100.0% Join 289 / 347 [1 x 1, 1 pairs] \n",
      "08:27 2.4Gb   100.0% Join 290 / 347 [2 x 3, 6 pairs]\n",
      "08:27 2.4Gb   100.0% Join 291 / 347 [5 x 11, 55 pairs]\n",
      "08:27 2.4Gb   100.0% Join 292 / 347 [5 x 16, 80 pairs]\n",
      "08:27 2.4Gb   100.0% Join 293 / 347 [2 x 21, 42 pairs]\n",
      "08:27 2.4Gb   100.0% Join 294 / 347 [2 x 2, 4 pairs]  \n",
      "08:27 2.4Gb   100.0% Join 295 / 347 [4 x 23, 92 pairs]\n",
      "08:27 2.4Gb   100.0% Join 296 / 347 [12 x 27, 324 pairs]\n",
      "08:27 2.4Gb   100.0% Join 297 / 347 [15 x 17, 255 pairs]\n",
      "08:27 2.4Gb   100.0% Join 298 / 347 [3 x 6, 18 pairs]   \n",
      "08:27 2.4Gb   100.0% Join 299 / 347 [7 x 8, 56 pairs]\n",
      "08:27 2.4Gb   100.0% Join 300 / 347 [9 x 15, 135 pairs]\n",
      "08:28 2.4Gb   100.0% Join 301 / 347 [24 x 197, 2000 pairs]\n",
      "08:28 2.4Gb   100.0% Join 302 / 347 [3 x 221, 663 pairs]  \n",
      "08:28 2.4Gb   100.0% Join 303 / 347 [8 x 224, 1792 pairs]\n",
      "08:28 2.4Gb   100.0% Join 304 / 347 [32 x 232, 2000 pairs]\n",
      "08:28 2.4Gb   100.0% Join 305 / 347 [1 x 264, 264 pairs]  \n",
      "08:28 2.4Gb   100.0% Join 306 / 347 [2 x 265, 530 pairs]\n",
      "08:28 2.4Gb   100.0% Join 307 / 347 [1 x 57, 57 pairs]  \n",
      "08:29 2.4Gb   100.0% Join 308 / 347 [58 x 267, 2000 pairs]\n",
      "08:29 2.4Gb   100.0% Join 309 / 347 [1 x 325, 325 pairs]  \n",
      "08:29 2.4Gb   100.0% Join 310 / 347 [39 x 326, 2000 pairs]\n",
      "08:29 2.4Gb   100.0% Join 311 / 347 [1 x 365, 365 pairs]  \n",
      "08:29 2.4Gb   100.0% Join 312 / 347 [64 x 366, 2000 pairs]\n",
      "08:30 2.4Gb   100.0% Join 313 / 347 [97 x 430, 2000 pairs]\n",
      "08:30 2.4Gb   100.0% Join 314 / 347 [1 x 527, 527 pairs]  \n",
      "08:30 2.4Gb   100.0% Join 315 / 347 [4 x 528, 2112 pairs]\n",
      "08:30 2.4Gb   100.0% Join 316 / 347 [1 x 532, 532 pairs] \n",
      "08:30 2.4Gb   100.0% Join 317 / 347 [1 x 533, 533 pairs]\n",
      "08:30 2.4Gb   100.0% Join 318 / 347 [2 x 534, 1068 pairs]\n",
      "08:31 2.4Gb   100.0% Join 319 / 347 [1 x 536, 536 pairs] \n",
      "08:31 2.4Gb   100.0% Join 320 / 347 [2 x 537, 1074 pairs]\n",
      "08:31 2.4Gb   100.0% Join 321 / 347 [1 x 539, 539 pairs] \n",
      "08:31 2.4Gb   100.0% Join 322 / 347 [1 x 540, 540 pairs]\n",
      "08:31 2.4Gb   100.0% Join 323 / 347 [1 x 541, 541 pairs]\n",
      "08:31 2.4Gb   100.0% Join 324 / 347 [14 x 542, 2000 pairs]\n",
      "08:32 2.4Gb   100.0% Join 325 / 347 [1 x 556, 556 pairs]  \n",
      "08:32 2.4Gb   100.0% Join 326 / 347 [2 x 557, 1114 pairs]\n",
      "08:32 2.4Gb   100.0% Join 327 / 347 [1 x 559, 559 pairs] \n",
      "08:32 2.4Gb   100.0% Join 328 / 347 [1 x 560, 560 pairs]\n",
      "08:32 2.4Gb   100.0% Join 329 / 347 [3 x 561, 1683 pairs]\n",
      "08:32 2.4Gb   100.0% Join 330 / 347 [1 x 564, 564 pairs] \n",
      "08:32 2.4Gb   100.0% Join 331 / 347 [2 x 565, 1130 pairs]\n",
      "08:33 2.4Gb   100.0% Join 332 / 347 [2 x 567, 1134 pairs]\n",
      "08:33 2.4Gb   100.0% Join 333 / 347 [1 x 569, 569 pairs] \n",
      "08:33 2.4Gb   100.0% Join 334 / 347 [1 x 570, 570 pairs]\n",
      "08:33 2.4Gb   100.0% Join 335 / 347 [2 x 571, 1142 pairs]\n",
      "08:33 2.4Gb   100.0% Join 336 / 347 [1 x 573, 573 pairs] \n",
      "08:33 2.4Gb   100.0% Join 337 / 347 [1 x 574, 574 pairs]\n",
      "08:33 2.4Gb   100.0% Join 338 / 347 [1 x 575, 575 pairs]\n",
      "08:34 2.4Gb   100.0% Join 339 / 347 [1 x 576, 576 pairs]\n",
      "08:34 2.4Gb   100.0% Join 340 / 347 [10 x 577, 2000 pairs]\n",
      "08:34 2.4Gb   100.0% Join 341 / 347 [587 x 1976, 2000 pairs]\n",
      "08:34 2.4Gb   100.0% Join 342 / 347 [1 x 2563, 2563 pairs]  \n",
      "08:35 2.4Gb   100.0% Join 343 / 347 [12 x 2564, 2000 pairs]\n",
      "08:35 2.4Gb   100.0% Join 344 / 347 [2 x 2576, 2000 pairs] \n",
      "08:36 2.4Gb   100.0% Join 345 / 347 [5 x 2578, 2000 pairs]\n",
      "08:36 2.4Gb   100.0% Join 346 / 347 [914 x 2583, 2000 pairs]\n",
      "08:36 2.4Gb   100.0% Join 347 / 347 [1 x 3497, 2000 pairs]  \n",
      "Inserting 5379 dupes... done.                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All position after alignment: (961, 8)\n",
      "After drop - positon: (135, 8)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "rm: cannot remove 'data/Bacdive/Designed_primers/Soil/designed_primers/v1v3/forward_pri_cov_temp.txt': No such file or directory\n",
      "primer_match: fasta_io.t:733: void StreamedFastaFile<T, H>::pos(FILE_POSITION_TYPE) [with T = MapFileChars; H = Lazy_Header_SI; FILE_POSITION_TYPE = long int]: Assertion `0' failed.\n",
      "Aborted (core dumped)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maybe unusual genus name: candidatus protochlamydia.\n",
      "Maybe unusual genus name: candidatus solibacter.\n",
      "Maybe unusual genus name: halalkalibacterium (ex joshi et al. 2022).\n",
      "Have -1. 1 -1\n",
      "Get 13496 records.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "muscle 5.1.linux64 [12f0e2]  1056Gb RAM, 192 cores\n",
      "Built Jan 13 2022 23:17:13\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 13496 seqs, length avg 128 max 134\n",
      "\n",
      "00:00 12Mb    100.0% Derep 6507 uniques, 6988 dupes\n",
      "00:00 14Mb   CPU has 192 cores, running 32 threads                   \n",
      "01:19 16Mb     96.5% UCLUST 6508 seqs EE<0.01, 785 centroids, 5496 members\r"
     ]
    }
   ],
   "source": [
    "vv_type_df = pd.read_csv('data/vvs_ontarget_marker.csv', index_col=0)\n",
    "vv_type_df['vv'] = list(vv_type_df.index)\n",
    "vv_type_df = pd.DataFrame(vv_type_df, columns=['vv', 'marker'])\n",
    "mid_size_vvs = list(vv_type_df[vv_type_df['marker'] == 'M']['vv'])\n",
    "\n",
    "# 设计引物用的参数\n",
    "mid_size_vvs = ';'.join(mid_size_vvs) # 输入的candidate vv区域\n",
    "num_every_spe = 50\n",
    "extend_bp_num = 50\n",
    "primer_lens = '20,20'\n",
    "k_num=3\n",
    "\n",
    "pcr_root = os.path.join(bacdive_root, 'PCR_validation')\n",
    "bacdive_pri_root = os.path.join(bacdive_root, 'Designed_primers')\n",
    "os.makedirs(bacdive_pri_root, exist_ok=True)\n",
    "    \n",
    "# read bacdive tsv file\n",
    "bacdive_csv = os.path.join(bacdive_root, 'Env_micro_list_genus.tsv')\n",
    "environment_info_df = pd.read_csv(bacdive_csv, sep='\\t', index_col=0)\n",
    "environment_info_df['Micro_spe'] = environment_info_df['Micro_spe'].apply(eval)\n",
    "environment_info_df['Environment_name'] = list(environment_info_df.index)\n",
    "\n",
    "# simple filter\n",
    "environment_info_df = environment_info_df[(environment_info_df['Genus_num'] >= 1)]\n",
    "environment_info_df.reset_index(inplace=True, drop=True)\n",
    "\n",
    "envi_primers_path = []\n",
    "design_failed_num = 0\n",
    "    \n",
    "# start design\n",
    "def create_abunTab(core_microbiota = [], file_to = ''):\n",
    "    df_abun = {'Genus': core_microbiota, 'err_a': len(core_microbiota) * [0.1], 'err_b': len(core_microbiota) * [0.1]}\n",
    "    df_abun = pd.DataFrame(df_abun)\n",
    "    df_abun.to_csv(file_to, index=False)\n",
    "    \n",
    "for i in tqdm(range(environment_info_df.shape[0])):\n",
    "    info_t = {}\n",
    "    microbiota_target = environment_info_df['Environment_name'][i]\n",
    "    info_t['ori_name'] = microbiota_target\n",
    "    microbiota_target = microbiota_target.strip().strip('#')\n",
    "    microbiota_target = '_'.join(microbiota_target.split())\n",
    "    microbiota_target = microbiota_target.replace('/', '_or_')\n",
    "    microbiota_target = microbiota_target.replace('(', '').replace(')', '') # special character that cannot be handled\n",
    "    \n",
    "    core_microbiota = environment_info_df['Micro_spe'][i]\n",
    "    out_dir = os.path.join(bacdive_pri_root, microbiota_target)\n",
    "    input_csv = f'{out_dir}/abunTab.csv'\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    create_abunTab(core_microbiota, f'{out_dir}/abunTab.csv')\n",
    "    \n",
    "    # to run DEcoPrimer\n",
    "    bash_file = f'{out_dir}/run_DEcoPrimer.sh'\n",
    "    with open(bash_file, 'w') as f:\n",
    "        try:\n",
    "            cmd1 = f\"python DEcoPrimer.py --input {input_csv} --out_root {out_dir} --input_type 'genera_profiling' --target_vs '{mid_size_vvs}' --extend_bp_num {extend_bp_num} --num_every_spe {num_every_spe} --NGS_mode Single_end\"\n",
    "            os.system(cmd1)\n",
    "            cmd2 = f\"python Insilico_eva_primers.py --envi_forEva {out_dir} --primers_forEva '{out_dir};' --target_vs '{mid_size_vvs}' --K {k_num} --output {pcr_root}/{microbiota_target}\"\n",
    "            os.system(cmd2)\n",
    "            cmd3 = f\"python Screen_best_PP.py --pcr_dir {pcr_root}/{microbiota_target} --rk_by Genus_accuracy\"\n",
    "            os.system(cmd3)\n",
    "        except:\n",
    "            design_failed_num += 1\n",
    "        f.write(cmd1, cmd2, cmd3)\n",
    "        \n",
    "    \n",
    "    # record some metainfo\n",
    "    info_t['clear_name'] = microbiota_target\n",
    "    envi_primers_path.append(info_t)\n",
    "\n",
    "\n",
    "print(f'Env design failed: {design_failed_num}.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "对bacdive结果分析并生成文章结果图"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/DataProcess_and_InsilicoEvaluation.ipynb 单元格 15\u001b[0m line \u001b[0;36m2\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2B7b22686f73744e616d65223a224e65775f64656c6c227d/data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/DataProcess_and_InsilicoEvaluation.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m## 获取每个环境的设计结果+整理metainfo+预处理数据\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell://ssh-remote%2B7b22686f73744e616d65223a224e65775f64656c6c227d/data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/DataProcess_and_InsilicoEvaluation.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m bacdive_ori_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\u001b[39m'\u001b[39m\u001b[39mdata/Bacdive/isolation_sources_bacdive_2022-10-07.csv\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2B7b22686f73744e616d65223a224e65775f64656c6c227d/data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/DataProcess_and_InsilicoEvaluation.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m bacdive_ori_df \u001b[39m=\u001b[39m bacdive_ori_df\u001b[39m.\u001b[39mdropna(subset\u001b[39m=\u001b[39m[\u001b[39m'\u001b[39m\u001b[39mCategory 3\u001b[39m\u001b[39m'\u001b[39m])\u001b[39m.\u001b[39mreset_index(drop\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m      <a href='vscode-notebook-cell://ssh-remote%2B7b22686f73744e616d65223a224e65775f64656c6c227d/data1/hyzhang/Projects/EcoPrimer_git/DeepEcoPrimer_v2/DataProcess_and_InsilicoEvaluation.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3'>4</a>\u001b[0m bacdive_ori_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(bacdive_ori_df, columns\u001b[39m=\u001b[39m[\u001b[39m'\u001b[39m\u001b[39mCategory 1\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mCategory 2\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mCategory 3\u001b[39m\u001b[39m'\u001b[39m])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "## 获取每个环境的设计结果+整理metainfo+预处理数据\n",
    "bacdive_ori_df = pd.read_csv('data/Bacdive/isolation_sources_bacdive_2022-10-07.csv')\n",
    "bacdive_ori_df = bacdive_ori_df.dropna(subset=['Category 3']).reset_index(drop=True)\n",
    "bacdive_ori_df = pd.DataFrame(bacdive_ori_df, columns=['Category 1', 'Category 2', 'Category 3'])\n",
    "bacdive_ori_df.drop_duplicates(inplace=True)\n",
    "bacdive_ori_df.reset_index(drop=True,inplace=True)\n",
    "\n",
    "bacdive_to_df = pd.read_csv('data/Bacdive/Results/Env_micro_bacdive.csv')\n",
    "bacdive_to_df = pd.merge(bacdive_to_df, bacdive_ori_df, how='left', left_on='Environment_name', right_on='Category 3').drop(columns=['Category 3'])\n",
    "print(bacdive_to_df['Category 2'].value_counts())\n",
    "bacdive_to_df.to_csv('data/Bacdive/Results/Env_micro_bacdive_mergeInfo.csv', index=False) # 保存metainfo table\n",
    "\n",
    "# 获取引物设计结果路径和信息\n",
    "bacdive_root = 'data/Bacdive'\n",
    "evi_list = os.listdir(bacdive_root)\n",
    "evi_list = [x for x in evi_list if os.path.isdir(os.path.join(bacdive_root, x))]\n",
    "rank_by = 'Genus_accuracy' # 'pri_amp_eff' 'Genus_accuracy'\n",
    "vvs_midlen = ['v1v2', 'v1v3', 'v4v5', 'v3v4', 'v6v8', 'v5v7', 'v7v8', 'v4v4', 'v5v6']\n",
    "\n",
    "univer_primer_table = pd.read_excel('data/Bacdive/Results/primers_sequencing_bacdiveComp.xlsx')# 进行细致比较的时候会用到\n",
    "univer_primer_table = {x: f'{y}_{x}' for x, y in zip(univer_primer_table['v_region'], univer_primer_table['pri_nm'])}\n",
    "\n",
    "len(evi_list), univer_primer_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. 做的genus-level的均值的结果\n",
    "bacdive_metainfo = pd.read_csv(f'{bacdive_root}/Env_micro_bacdive_mergeInfo.csv')\n",
    "bacdive_metainfo.drop(columns=['Information', 'Micro_spe'], inplace=True)\n",
    "\n",
    "rk_by = 'Genus_accuracy'\n",
    "df_res_all_evi = pd.read_csv(f'{bacdive_root}/bacdive_designed_res_{rk_by}.csv')\n",
    "\n",
    "def get_vv_region(pri_str, ):\n",
    "    pt = 'v\\dv\\d'\n",
    "    vv_region = re.findall(pt, pri_str)[0]\n",
    "    if pri_str.endswith(vv_region):\n",
    "        pri_type = 'universal'\n",
    "    else:\n",
    "        pri_type = 'designed'\n",
    "    return vv_region, pri_type\n",
    "\n",
    "df_res_all_evi['VV_region'] = df_res_all_evi['Primer'].apply(lambda x:get_vv_region(x)[0])\n",
    "df_res_all_evi['Pri_type'] = df_res_all_evi['Primer'].apply(lambda x:get_vv_region(x)[1])\n",
    "df_res_all_evi = df_res_all_evi[df_res_all_evi['VV_region'].isin(vvs_midlen)] # 保证只在这4个v区域中间选择\n",
    "df_res_all_evi.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# 找每个evi的结果\n",
    "bacdive_best_primer = []\n",
    "for evi, df in df_res_all_evi.groupby('Environment'):\n",
    "    idx = df['Value'].idxmax()\n",
    "    uni_pri_res = {}\n",
    "    for uni_reg, uni_nm in univer_primer_table.items():\n",
    "        try:\n",
    "            v3v4_uni_acc = list(df[df['Primer'] == uni_nm]['Value'])[0]\n",
    "        except:\n",
    "            v3v4_uni_acc = 0.0\n",
    "        uni_pri_res[uni_nm] = v3v4_uni_acc\n",
    "    \n",
    "    info_t = {'Environment': evi, 'Primer': df.loc[idx, 'Primer'], 'Value': df.loc[idx, 'Value'], 'VV_region': df.loc[idx, 'VV_region'], 'Pri_type': df.loc[idx, 'Pri_type']}\n",
    "    info_t.update(uni_pri_res)\n",
    "    \n",
    "    # 看一下设计引物的最优\n",
    "    df_de = df[df['Pri_type'] == 'designed'].copy()\n",
    "    df_de.reset_index(inplace=True, drop=True)\n",
    "    idx = df_de['Value'].idxmax()\n",
    "    info_t.update({'Designed_best_val': df_de.loc[idx, 'Value'], 'Designed_best_primer': df_de.loc[idx, 'Primer'], 'Designed_best_vv_region': df_de.loc[idx, 'VV_region']})\n",
    "    \n",
    "    # 看一下通用引物的最优\n",
    "    df_un = df[df['Pri_type'] == 'universal'].copy()\n",
    "    df_un.reset_index(inplace=True, drop=True)\n",
    "    idx = df_un['Value'].idxmax()\n",
    "    info_t.update({'Universal_best_val': df_un.loc[idx, 'Value'], 'Universal_best_primer': df_un.loc[idx, 'Primer'], 'Universal_best_vv_region': df_un.loc[idx, 'VV_region']})\n",
    "    \n",
    "    bacdive_best_primer.append(info_t)\n",
    "    \n",
    "bacdive_best_primer = pd.DataFrame(bacdive_best_primer)\n",
    "bacdive_best_primer = pd.merge(bacdive_best_primer, bacdive_metainfo, how='left', left_on='Environment', right_on='Environment_name_processed')\n",
    "# print(bacdive_best_primer[bacdive_best_primer['Environment'].duplicated()])\n",
    "bacdive_best_primer['De_minus_bestUni'] = bacdive_best_primer['Designed_best_val'] - bacdive_best_primer['Universal_best_val']\n",
    "bacdive_best_primer.to_csv(f'results/bacdive/bacdive_best_primer_{rk_by}.csv', index=False)\n",
    "\n",
    "# 1. 画一个柱状图\n",
    "bacdive_best_primer = bacdive_best_primer[bacdive_best_primer['Genus_num'] > 10] # 按照target genus数目做一下筛选\n",
    "print(f'Total envi number: {bacdive_best_primer.shape}')\n",
    "bacdive_best_primer.reset_index(inplace=True, drop=True)\n",
    "bacdive_best_primer['bar_label'] = bacdive_best_primer['VV_region'] + '_' + bacdive_best_primer['Pri_type']\n",
    "bar_plot_info_t = dict(bacdive_best_primer['bar_label'].value_counts())\n",
    "bar_plot_info = {k: v for k, v in bar_plot_info_t.items() if 'desi' in k}\n",
    "bar_plot_info.update({k: v for k, v in bar_plot_info_t.items() if 'uni' in k})\n",
    "\n",
    "# plt.style.use('ggplot')\n",
    "plt.figure(figsize=(8, 15))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "x, y = list(bar_plot_info.keys()), list(bar_plot_info.values())\n",
    "x_val = list(range(len(y))) \n",
    "x_des_only = x_val[: -2] # 因为de包含所有的v区域，所以要去掉最后两个\n",
    "\n",
    "colors = {'Universal': '#8ECFC9', 'Designed': '#FA7F6F'}\n",
    "uni_des_gps = {'Universal':{'x': [], 'y': []}, 'Designed':{'x': [], 'y': []}}\n",
    "\n",
    "# 先放designed，再放universal\n",
    "x_for_uni = {'v4v5_universal': 1, 'v3v4_universal': 0}\n",
    "for i, label in enumerate(x):\n",
    "    if 'designed' in label:\n",
    "        uni_des_gps['Designed']['x'].append(x_val[i])\n",
    "        uni_des_gps['Designed']['y'].append(y[i])\n",
    "    else:\n",
    "        uni_des_gps['Universal']['x'].append(x_for_uni[label])   # x_val[i])\n",
    "        uni_des_gps['Universal']['y'].append(y[i])\n",
    "\n",
    "wid = 0.3\n",
    "for lb, idx in uni_des_gps.items():\n",
    "    if lb == 'Universal':\n",
    "        plt.bar(np.array(idx['x']) + wid / 2, np.array(idx['y']), width=wid, label=lb + ' primer pairs', color=colors[lb], )\n",
    "        for i_x in x_des_only:\n",
    "            try:\n",
    "                x_, y_ = idx['x'][i_x], idx['y'][i_x]\n",
    "            except:\n",
    "                x_, y_ = i_x, 0\n",
    "            plt.text(x_ + wid / 2, y_+1, y_,ha='center',fontsize=10)\n",
    "        \n",
    "    else:\n",
    "        plt.bar(np.array(idx['x']) - wid / 2, np.array(idx['y']), width=wid, label=lb + ' primer pairs', color=colors[lb], )\n",
    "        for i_x in x_des_only:\n",
    "            try:\n",
    "                x_, y_ = idx['x'][i_x], idx['y'][i_x]\n",
    "            except:\n",
    "                x_, y_ = i_x, 0\n",
    "            plt.text(x_ - wid / 2, y_+1, y_,ha='center',fontsize=10)\n",
    "\n",
    "x_labs = [x.split('_')[0] for x in x[: -2]]\n",
    "plt.xticks(x_val[: -2], [x.upper() for x in x_labs], fontsize=10, rotation=315)\n",
    "plt.xlabel('The selection of the best V-region')\n",
    "plt.ylabel('Number of environments')\n",
    "# plt.title('Best primer for each environment')\n",
    "plt.legend()\n",
    "plt.grid(axis='y', linestyle='--')\n",
    "# 设置边框不可见\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "# plt.tight_layout()\n",
    "plt.savefig(f'data/Bacdive/Results/bacdive_best_primer_{rk_by}.svg', dpi=800, bbox_inches='tight')\n",
    "\n",
    "# 2. 按照各个环境画堆叠柱状图\n",
    "evi_category = 'Category 2'\n",
    "Cate_index = bacdive_best_primer[evi_category].value_counts()\n",
    "Cate_index = list(Cate_index[Cate_index >= 8].index)\n",
    "Cate_index.remove('#Other')\n",
    "# vv_index = x\n",
    "vv_index = x_labs\n",
    "\n",
    "df_category_bar = []\n",
    "for cate_i, df in bacdive_best_primer.groupby(evi_category,):\n",
    "    if cate_i not in Cate_index:\n",
    "        continue\n",
    "    vv_cnt = dict(df['VV_region'].value_counts()) # bar_label用作备选方案\n",
    "    vv_cnt[evi_category] = cate_i\n",
    "    df_category_bar.append(vv_cnt)\n",
    "\n",
    "df_category_bar = pd.DataFrame(df_category_bar, columns = [evi_category] + vv_index)\n",
    "df_category_bar.fillna(0, inplace=True)\n",
    "df_category_bar.index = df_category_bar[evi_category]\n",
    "df_category_bar.drop(columns=[evi_category], inplace=True)\n",
    "\n",
    "# df_category_bar = df_category_bar.loc[Cate_index, :] # 按照环境数目的顺序来重新排序\n",
    "df_category_bar = df_category_bar / df_category_bar.sum(axis=1).values.reshape(-1, 1)\n",
    "df_category_bar['v3_v5_sum'] = df_category_bar.iloc[:, : 2].sum(axis=1) # v3v4 + v4v5去排序\n",
    "df_category_bar.sort_values(by = 'v3_v5_sum', inplace=True, ascending=False)\n",
    "\n",
    "plt.figure(figsize=(10, 20))\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "cmp = mpl.cm.get_cmap('Set3', len(vv_index))\n",
    "for vv_i in range(len(vv_index)):\n",
    "    plot_info = df_category_bar.iloc[:, vv_i]\n",
    "    if vv_i == 0:\n",
    "        plt.bar(np.arange(plot_info.shape[0]), plot_info, label=vv_index[vv_i].upper(), color= cmp(vv_i / len(vv_index)))\n",
    "    else:\n",
    "        plot_info_bottom = df_category_bar.iloc[:, : vv_i].sum(axis=1)\n",
    "        plt.bar(np.arange(plot_info.shape[0]), plot_info, bottom=plot_info_bottom, label=vv_index[vv_i].upper(), color= cmp(vv_i / len(vv_index)))\n",
    "\n",
    "xtick_cate_nm = list(df_category_bar.index)\n",
    "plt.xticks(np.arange(plot_info.shape[0]), xtick_cate_nm, fontsize=8, rotation=315, ha='left')\n",
    "# 调整刻度标签与坐标轴的距离\n",
    "plt.tick_params(axis='x', pad=1)  # pad 参数设置刻度标签与坐标轴的距离\n",
    "\n",
    "# 自定义 y 轴上的刻度标签为百分比形式\n",
    "plt.ylabel('Relative ratio (%)')\n",
    "def percentage_formatter(x, pos):\n",
    "    return f'{x*100:.0f}%'\n",
    "plt.gca().yaxis.set_major_formatter(FuncFormatter(percentage_formatter))\n",
    "\n",
    "\n",
    "# plt.tight_layout()\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "# plt.title('Best primer for each category.')\n",
    "# 设置边框不可见\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "plt.savefig(f'data/Bacdive/Results/bacdive_best_primer_category_{rk_by}.svg', dpi=800, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 用来进行v3v4细致比较的func\n",
    "def compDesign_v3v4(de_best_pri_ge_list, uni_target_pri_ge_list, ):\n",
    "    # 成对样本检验p-value\n",
    "    pval_paired_test = wilcoxon(de_best_pri_ge_list, uni_target_pri_ge_list, zero_method='wilcox', correction=False).pvalue\n",
    "    return pval_paired_test\n",
    "    \n",
    "    \n",
    "xlsx_write = pd.ExcelWriter(r'results/bacdive/detail.xlsx')\n",
    "# 3. 每个环境做细致的genus比较\n",
    "# 0521：细致比较每一个环境和v3v4的区别\n",
    "uni_comp_vregion = 'v3v4'\n",
    "\n",
    "rk_by = 'Genus_accuracy'\n",
    "bacdive_root = 'IN_bacdive_middle_v2'\n",
    "\n",
    "vv_uni_comp = univer_primer_table[uni_comp_vregion] #'338F_806R_v3v4'\n",
    "accDiff_cut_off = 0.05\n",
    "pval_cutoff = 0.05 # 画图的时候的pval cut off\n",
    "# 0619: 进行列联表检验的参数\n",
    "eff_ge_moreThan_seqNum = 10 # 只保留那些序列数大于一定数目的ge\n",
    "prevalent_rare_divNum = 100 # 用来划分prevalent和rare的genus数目\n",
    "\n",
    "ge_spectrum_comp = []\n",
    "ge_paired_test = []\n",
    "ge_corr_test = []\n",
    "for evi in list(bacdive_best_primer['Environment']):\n",
    "    evi_summary_df = pd.read_csv(os.path.join(bacdive_root, evi, 'detail_Genus_accuracy.csv'))\n",
    "    evi_summary_df.index = evi_summary_df['tax_name']\n",
    "    de_best_pri = list(bacdive_best_primer[bacdive_best_primer['Environment'] == evi]['Designed_best_primer'])[0]\n",
    "    try:\n",
    "        # 找到比较对象的ge-spectrum\n",
    "        de_best_pri_ge_list = evi_summary_df[de_best_pri]\n",
    "        uni_target_pri_ge_list = evi_summary_df[vv_uni_comp]\n",
    "        ge_representative_seqNum = evi_summary_df['tax_num']\n",
    "    except:\n",
    "        ge_spectrum_comp.append(None)\n",
    "        ge_paired_test.append(None)\n",
    "        continue\n",
    "    \n",
    "    # 1.计算成对样本检验情况；2.后续加入seq num列联表检验\n",
    "    pval_paired_test = compDesign_v3v4(de_best_pri_ge_list, uni_target_pri_ge_list)\n",
    "    ge_paired_test.append(pval_paired_test)\n",
    "    \n",
    "    # 细致看哪些ge超过了cut_off\n",
    "    evi_ge_spectrum = {}\n",
    "    a, b, c, d, e, f = 0, 0, 0, 0, 0, 0\n",
    "    \n",
    "    for i in range(len(de_best_pri_ge_list)):\n",
    "        # 去除那些只有较少representative序列的ge; TODO:试一下acc计算的时候，也去除\n",
    "        if ge_representative_seqNum[i] < eff_ge_moreThan_seqNum:\n",
    "            continue\n",
    "            \n",
    "        if de_best_pri_ge_list[i] - uni_target_pri_ge_list[i] > accDiff_cut_off:\n",
    "            # 显著优于uni primer\n",
    "            evi_ge_spectrum.update({de_best_pri_ge_list.index[i]: (de_best_pri_ge_list[i], uni_target_pri_ge_list[i])})\n",
    "            if ge_representative_seqNum[i] > prevalent_rare_divNum:\n",
    "                a += 1\n",
    "            else:\n",
    "                d += 1\n",
    "                \n",
    "        elif de_best_pri_ge_list[i] - uni_target_pri_ge_list[i] < -accDiff_cut_off:\n",
    "            # 显著更差于uni primer\n",
    "            if ge_representative_seqNum[i] > prevalent_rare_divNum:\n",
    "                b += 1\n",
    "            else:\n",
    "                e += 1\n",
    "        \n",
    "        else:\n",
    "            # 没有显著差异\n",
    "            if ge_representative_seqNum[i] > prevalent_rare_divNum:\n",
    "                c += 1\n",
    "            else:\n",
    "                f += 1\n",
    "    try:\n",
    "        stat_data = [[a, b, c],\n",
    "                    [d, e, f]]\n",
    "        chi2, p, dof, expected = chi2_contingency(stat_data)\n",
    "    except:\n",
    "        p = -1.0\n",
    "        stat_data = []\n",
    "    \n",
    "    ge_corr_test.append((p, stat_data))\n",
    "    ge_spectrum_comp.append(evi_ge_spectrum)\n",
    "    \n",
    "bacdive_best_primer[f'Spectrum_comp_{vv_uni_comp}'] = ge_spectrum_comp\n",
    "bacdive_best_primer[f'Paired_test_{vv_uni_comp}'] = ge_paired_test\n",
    "bacdive_best_primer[f'Corr_test_{vv_uni_comp}_{prevalent_rare_divNum}_{eff_ge_moreThan_seqNum}'] = ge_corr_test\n",
    "\n",
    "# 4. 画按照genus-num递减的环境散点图\n",
    "bacdive_best_primer.sort_values(by='Genus_num', inplace=True, ascending=False)\n",
    "bacdive_best_primer.reset_index(inplace=True, drop=True)\n",
    "# plt.style.use('ggplot')\n",
    "plt.figure()\n",
    "# bacdive_best_primer.iloc[:180, :].plot.scatter(x='Genus_num', y='Value', figsize=(8, 6), s=50, c='DarkBlue', )\n",
    "plt.plot(np.log(bacdive_best_primer['Genus_num']), bacdive_best_primer['Value'], 'o', markersize=3)\n",
    "plt.title('Best primer of environment with different genus number')\n",
    "plt.savefig(f'results/bacdive/bacdive_best_primer_genusNum.pdf', dpi=300)\n",
    "\n",
    "# 细致看看比v3v4更好的环境\n",
    "bacdive_best_primer[f'Design_ninus_{uni_comp_vregion}'] = bacdive_best_primer['Designed_best_val'] - bacdive_best_primer[vv_uni_comp]\n",
    "\n",
    "print('Degree of designed primer better than universal primer: \\n', bacdive_best_primer[f'Designed_best_val'].describe(), bacdive_best_primer[vv_uni_comp].describe()) # bacdive_best_primer[bacdive_best_primer[vv_uni_comp] > 1e-5][f'Design_ninus_{uni_comp_vregion}'].describe()\n",
    "bacdive_best_primer_greaterThan_v3v4 = bacdive_best_primer[(bacdive_best_primer[f'Design_ninus_{uni_comp_vregion}'] >= accDiff_cut_off) & (bacdive_best_primer[vv_uni_comp] > 1e-5) & (bacdive_best_primer[f'Paired_test_{vv_uni_comp}'] < pval_cutoff)]\n",
    "bacdive_best_primer_greaterThan_v3v4.reset_index(inplace=True, drop=True)\n",
    "bacdive_best_primer_greaterThan_v3v4.to_excel(xlsx_write, sheet_name=f'Greater_than_{uni_comp_vregion}_{accDiff_cut_off}_{pval_cutoff}', index=False)\n",
    "print('Degree of designed primer better than universal primer: \\n', bacdive_best_primer_greaterThan_v3v4['Design_ninus_v3v4'].describe())\n",
    "\n",
    "# 和v3v4比较的时候画一个成对样本检验p-value和acc大于的火山图\n",
    "df = bacdive_best_primer.copy()\n",
    "df = df[df[vv_uni_comp] > 1e-5] # 排除掉没有v3v4结果的\n",
    "df.reset_index(inplace=True, drop=True)\n",
    "df[f'Paired_test_{vv_uni_comp}']=df[f'Paired_test_{vv_uni_comp}'].apply(lambda x:-np.log10(x+1e-14))\n",
    "\n",
    "plt.figure(figsize=(12,10))\n",
    "plt.scatter(x=df[f'Design_ninus_{uni_comp_vregion}'],y=df[f'Paired_test_{vv_uni_comp}'],s=10,label=\"Not significant\",color=\"grey\")\n",
    "# highlight down- or up- regulated genes\n",
    "down = df[(df[f'Design_ninus_{uni_comp_vregion}']<=-accDiff_cut_off) & (df[f'Paired_test_{vv_uni_comp}']>=-math.log10(pval_cutoff))]\n",
    "up = df[(df[f'Design_ninus_{uni_comp_vregion}']>=accDiff_cut_off) & (df[f'Paired_test_{vv_uni_comp}']>=-math.log10(pval_cutoff))]\n",
    "\n",
    "plt.scatter(x=down[f'Design_ninus_{uni_comp_vregion}'],y=down[f'Paired_test_{vv_uni_comp}'],s=15,label=\"16sDeepEcoPrimer is worse\",color=\"blue\")\n",
    "plt.scatter(x=up[f'Design_ninus_{uni_comp_vregion}'],y=up[f'Paired_test_{vv_uni_comp}'],s=15,label=\"16sDeepEcoPrimer is better\",color=\"red\")\n",
    "\n",
    "texts=[]\n",
    "for i,r in down.iterrows():\n",
    "    texts.append(plt.text(x=r[f'Design_ninus_{uni_comp_vregion}'],y=r[f'Paired_test_{vv_uni_comp}'],s=str(r['Environment']))) \n",
    "for i,r in up.iterrows():\n",
    "    texts.append(plt.text(x=r[f'Design_ninus_{uni_comp_vregion}'],y=r[f'Paired_test_{vv_uni_comp}'],s=str(r['Environment']))) \n",
    "\n",
    "if len(texts) > 0:\n",
    "    adjust_text(texts, arrowprops=dict(arrowstyle='-', color='black'))\n",
    "\n",
    "\n",
    "plt.xlabel(f\"Classification accuracy of designed primer minus that of 338F_806R\", fontsize=15)\n",
    "plt.ylabel(\"-logFDR\", fontsize=15)\n",
    "# 自定义 x轴上的刻度标签为百分比形式\n",
    "def percentage_formatter(x, pos):\n",
    "    return f'{x*100:.0f}%'\n",
    "plt.gca().xaxis.set_major_formatter(FuncFormatter(percentage_formatter))\n",
    "\n",
    "# plt.title(f'Difference of classification accuracy between designed primer and 338F_806R') #  {vv_uni_comp}\n",
    "plt.axvline(-accDiff_cut_off,color=\"grey\",linestyle=\"--\")\n",
    "plt.axvline(accDiff_cut_off,color=\"grey\",linestyle=\"--\")\n",
    "plt.axhline(-math.log10(pval_cutoff),color=\"grey\",linestyle=\"--\")\n",
    "\n",
    "plt.xticks(fontsize=15)\n",
    "plt.yticks(fontsize=15)\n",
    "plt.legend(fontsize='large')\n",
    "plt.savefig(f'results/bacdive/bacdive_compWith_{vv_uni_comp}.svg', dpi=600, bbox_inches='tight')\n",
    "\n",
    "\n",
    "# 5. 看看当最优是universal的时候，设计的效果如何\n",
    "bacdive_best_primer['Uni_ninus_designed'] = bacdive_best_primer['Designed_best_val'] - bacdive_best_primer['Value']\n",
    "bacdive_best_primer_uni_better = bacdive_best_primer[bacdive_best_primer['Pri_type'] == 'universal']\n",
    "print('Info of environment when universal primer is better than designed primer: \\n', bacdive_best_primer_uni_better['Uni_ninus_designed'].describe())\n",
    "print(bacdive_best_primer_uni_better['Designed_best_val'].describe(), bacdive_best_primer_uni_better['Universal_best_val'].describe())\n",
    "\n",
    "plt.figure()\n",
    "\n",
    "# 6. 看看当最优是design的时候，能够优秀多少\n",
    "bacdive_best_primer_de_better = bacdive_best_primer[bacdive_best_primer['Pri_type'] == 'designed']\n",
    "print('Info of environment when designed primer is better: \\n', bacdive_best_primer_de_better['De_minus_bestUni'].describe())\n",
    "print(bacdive_best_primer_de_better['Designed_best_val'].describe(), bacdive_best_primer_de_better['Universal_best_val'].describe())\n",
    "\n",
    "ind_plot = ['Designed_best_val', 'Value']\n",
    "for ind in ind_plot:\n",
    "    plt.plot(bacdive_best_primer_uni_better[ind], label=ind)\n",
    "plt.legend()\n",
    "\n",
    "xlsx_write.close()\n",
    "\n",
    "## primer best df中新加入了一些指标，重新保存\n",
    "bacdive_best_primer.to_csv(f'results/bacdive/bacdive_best_primer_{rk_by}_bigEvi_{vv_uni_comp}_Comp.csv', index=False)\n",
    "\n",
    "bacdive_best_primer_uni_better.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. 看看不同categories的\n",
    "comp_pris = ['27F_338R_v1v2', '27F_533R_v1v3', '515F_907R_v4v5', '338F_806R_v3v4', '515F_806R_v4v4', '969F_1406R_v6v8', 'Designed_best_val']\n",
    "evi_category = 'Category 2'\n",
    "Cate_index = bacdive_best_primer[evi_category].value_counts()\n",
    "Cate_index = list(Cate_index[Cate_index >= 8].index)\n",
    "Cate_index.remove('#Other')\n",
    "\n",
    "vv_index = x\n",
    "vv_index.sort()\n",
    "\n",
    "df_category_std = {\n",
    "    'Group':[],\n",
    "    'Subgroup':[],\n",
    "    'Value':[]\n",
    "    }\n",
    "for cate_i, df in bacdive_best_primer.groupby(evi_category,):\n",
    "    if cate_i not in Cate_index:\n",
    "        continue\n",
    "    gp1 = df['De_minus_bestUni'].values\n",
    "    gp2 = df[comp_pris].std(axis=1).values\n",
    "    gp1 = []\n",
    "    \n",
    "    df_category_std['Group'] += [cate_i] * (len(gp1) + len(gp2))\n",
    "    df_category_std['Subgroup'] += ['diff_1'] * len(gp1) + ['diff_2'] * len(gp2)\n",
    "    df_category_std['Value'] += list(gp1) + list(gp2)\n",
    "df_category_std = pd.DataFrame(df_category_std)\n",
    "\n",
    "# 计算每个组别的均值并排序\n",
    "mean_by_group = df_category_std.groupby('Group')['Value'].mean().sort_values()\n",
    "# 根据均值排序后的组别顺序，对数据框进行排序\n",
    "data_sorted = df_category_std.sort_values(by='Group', key=lambda x: x.map(mean_by_group))\n",
    "\n",
    "# 设置顶级科学期刊配色方案\n",
    "color_palette = sns.color_palette(\"husl\", 3)\n",
    "\n",
    "# 设置绘图风格为 'ticks'\n",
    "sns.set_style(\"ticks\")\n",
    "\n",
    "# 绘制双列的箱线图\n",
    "plt.figure(figsize=(12, 9))  # 设置图形大小\n",
    "sns.boxplot(x='Group', y='Value', data=data_sorted, width=0.45, color=color_palette[2])  # 绘制箱线图\n",
    "\n",
    "# 添加标题和坐标轴标签\n",
    "# plt.title('Primer bias in different categories')\n",
    "plt.ylabel('Primer bias (%)', fontsize=20)\n",
    "plt.xlabel('')\n",
    "plt.xticks(fontsize=20, rotation=315, ha='left')\n",
    "plt.yticks(fontsize=20)\n",
    "# 调整刻度标签与坐标轴的距离\n",
    "plt.tick_params(axis='x', pad=1)  # pad 参数设置刻度标签与坐标轴的距离\n",
    "plt.grid(axis='y', linestyle='--')\n",
    "\n",
    "# 自定义 y 轴上的刻度标签为百分比形式\n",
    "def percentage_formatter(x, pos):\n",
    "    return f'{x*100:.0f}%'\n",
    "plt.gca().yaxis.set_major_formatter(FuncFormatter(percentage_formatter))\n",
    "\n",
    "# 去除顶部和右侧的边框线\n",
    "sns.despine()\n",
    "\n",
    "plt.savefig(f'data/Bacdive/Results/Primer_bias.svg', dpi=600, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8. 看一下各个环境设计出的引物具体target位点，尝试解释为何我们的引物更好\n",
    "\n",
    "## 0630：找各个primer binding site的序列\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "bacdive_best_primer_metadf = pd.read_csv('results/bacdive/bacdive_best_primer_Genus_accuracy_bigEvi_338F_806R_v3v4_Comp.csv')\n",
    "bac_pri_root = 'bacdiveEvi_primers'\n",
    "\n",
    "def find_opt_pri_seq(evi_df):\n",
    "    opt_primer_infos = []\n",
    "    for evi_nm, pri_nm in zip(list(evi_df['Environment']), list(evi_df['Designed_best_primer'])):\n",
    "        # 首先提取meta信息\n",
    "        pri_nm += 'r'\n",
    "        vv, fr = pri_nm.split('_')[-2], pri_nm.split('_')[-1]\n",
    "        evi_root = os.path.join(bac_pri_root, evi_nm, vv)\n",
    "        f_num, r_num = int(fr.strip('r').split('f')[0]), int(fr.strip('r').split('f')[1])\n",
    "        \n",
    "        # 从csv文件中找到primer的详细信息\n",
    "        f_pri_df, r_pri_df = pd.read_csv(os.path.join(evi_root, f'forward_primer.csv')), pd.read_csv(os.path.join(evi_root, f'reverse_primer.csv'))\n",
    "        f_pri_df.sort_values(by='cov_num_k_3', ascending=False, inplace=True)\n",
    "        f_pri_df.reset_index(inplace=True, drop = True)\n",
    "        r_pri_df.sort_values(by='cov_num_k_3', ascending=False, inplace=True)\n",
    "        r_pri_df.reset_index(inplace=True, drop = True)\n",
    "\n",
    "        # 整合optimal primer信息\n",
    "        opt_primer_info_head = ['primer_atgc', 'primer_degenum', 'GC_ratio', 'Tm']\n",
    "        opt_primer_info_tab = {}\n",
    "        for opt_pri_if in opt_primer_info_head:\n",
    "            f_pri, r_pri = f_pri_df.loc[f_num, opt_pri_if], r_pri_df.loc[r_num, opt_pri_if]\n",
    "            opt_primer_info_tab[opt_pri_if + '_f'] = f_pri\n",
    "            opt_primer_info_tab[opt_pri_if + '_r'] = r_pri\n",
    "\n",
    "        opt_primer_infos.append(opt_primer_info_tab)\n",
    "    evi_df['opt_primer_info'] = opt_primer_infos\n",
    "    return evi_df\n",
    "\n",
    "bacdive_best_primer_metadf = find_opt_pri_seq(bacdive_best_primer_metadf)\n",
    "bacdive_best_primer_metadf.to_csv('results/bacdive/bacdive_optPrimerInfo.csv', index=False) # 保存来看看\n",
    "\n",
    "# 找到v3v4最优的那些设计引物\n",
    "design_pri_targetVregion = 'v3v4'\n",
    "bacdive_designed_primer_thisVregion = bacdive_best_primer_metadf[(bacdive_best_primer_metadf['Designed_best_vv_region'] == design_pri_targetVregion) & (bacdive_best_primer_metadf['Pri_type'] == 'designed')]\n",
    "bacdive_designed_primer_thisVregion = pd.DataFrame(bacdive_designed_primer_thisVregion, columns=['Designed_best_primer', 'Environment', 'opt_primer_info'])\n",
    "bacdive_designed_primer_thisVregion.reset_index(inplace=True, drop=True)\n",
    "\n",
    "univer_primer_table_ori = pd.read_excel('ref_database/universal_primers/primers_sequencing_bacdiveComp.xlsx')# 进行细致比较的时候会用到\n",
    "univer_primer_table_ori = univer_primer_table_ori[univer_primer_table_ori['v_region'] == design_pri_targetVregion].reset_index(drop=True)\n",
    "uni_pri_atgcs = {'primer_atgc_f':univer_primer_table_ori['forward_seq'][0].strip(\"5'- \").strip(\"-3'\"), 'primer_atgc_r':univer_primer_table_ori['reverse_seq'][0].strip(\"5'- \").strip(\"-3'\")}\n",
    "\n",
    "\n",
    "# 开始比对到ecoli上\n",
    "# 1. 先建立数据库\n",
    "ecoli_ref = 'ref_database/ecoli_K12_MG1655.fna'\n",
    "db = ecoli_ref.split(\".\")[0]\n",
    "cmd_mkDatabase = f'makeblastdb -in {ecoli_ref} -dbtype nucl -out {db}'\n",
    "os.system(cmd_mkDatabase)\n",
    "\n",
    "def blast_cmd(query_primer=\"\", db=\"\", out_path = ''):\n",
    "    cmd = f\"blastn -db {db} -query \"\n",
    "    cmd += query_primer\n",
    "    cmd += \" -outfmt 7 \"\n",
    "    cmd += \"-out \"\n",
    "    cmd += out_path\n",
    "    cmd += \" -task blastn-short -word_size 4 -evalue 1 -max_target_seqs 1000000\"\n",
    "    \n",
    "    os.system(cmd)\n",
    "    \n",
    "def parse_blastTxt(blast_txt = '/data3/hyzhang/ont/16s_RNA_seg/res/blast_res/query.txt'):\n",
    "    with open(blast_txt, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    index_line = lines[3]\n",
    "    index_line = index_line.split(': ')[-1]\n",
    "    index_line = index_line.split(',')\n",
    "    index_line = [x.strip(' ') for x in index_line]\n",
    "    index_line = ['_'.join(x.split()) for x in index_line]\n",
    "\n",
    "    lines = [x.strip('\\n').split('\\t') for x in lines if not x.startswith('#')]\n",
    "    \n",
    "    blast_df = pd.DataFrame(lines)\n",
    "    blast_df.columns = index_line\n",
    "    return blast_df\n",
    "\n",
    "# 2. 每个环境比对\n",
    "temp_dir = 'temp_align'\n",
    "df_evi_primer_ecoli = []\n",
    "for evi_i in range(bacdive_designed_primer_thisVregion.shape[0]):\n",
    "    evi_nm = bacdive_designed_primer_thisVregion.loc[evi_i, 'Environment']\n",
    "    f_pri, r_pri = bacdive_designed_primer_thisVregion.loc[evi_i, 'opt_primer_info']['primer_atgc_f'], bacdive_designed_primer_thisVregion.loc[evi_i, 'opt_primer_info']['primer_atgc_r']\n",
    "    records = [\n",
    "        SeqRecord(seq=Seq(f_pri), id='forward', description= ''),\n",
    "        SeqRecord(seq=Seq(r_pri), id='reverse', description= '')\n",
    "    ]\n",
    "\n",
    "    with open(f'./{temp_dir}/temp_al.fna', 'w') as f:\n",
    "        SeqIO.write(records, f, 'fasta')\n",
    "        \n",
    "    blast_cmd(query_primer=f'./{temp_dir}/temp_al.fna', db=db, out_path=f\"./{temp_dir}/{evi_nm}.fna\")\n",
    "    blast_df = parse_blastTxt(blast_txt=f\"./{temp_dir}/{evi_nm}.fna\")\n",
    "    \n",
    "    info_t = {'Environment': evi_nm}\n",
    "    for pri_ty, pri_atgc in zip(['forward', 'reverse'], [f_pri, r_pri]):\n",
    "        info_t[pri_ty + '_seq'] = pri_atgc\n",
    "        df_ = blast_df[blast_df['query_acc.ver'] == pri_ty].reset_index(drop=True)\n",
    "        df_.sort_values(by='evalue', inplace=True, ascending=True)\n",
    "        info_t[pri_ty + '_bitscore'] = df_.loc[0, 'bit_score']\n",
    "        if pri_ty == 'forward':\n",
    "            info_t[pri_ty + '_start'] = int(df_.loc[0, 's._start']) - (int(df_.loc[0, 'q._start']) - 1)\n",
    "        else:\n",
    "            info_t[pri_ty + '_start'] = int(df_.loc[0, 's._start']) + (int(df_.loc[0, 'q._start']) - 1)\n",
    "    df_evi_primer_ecoli.append(info_t)\n",
    "\n",
    "# 3. uni引物结果\n",
    "evi_nm = f'universal_{design_pri_targetVregion}'\n",
    "f_pri, r_pri = uni_pri_atgcs['primer_atgc_f'], uni_pri_atgcs['primer_atgc_r']\n",
    "records = [\n",
    "    SeqRecord(seq=Seq(f_pri), id='forward', description= ''),\n",
    "    SeqRecord(seq=Seq(r_pri), id='reverse', description= '')\n",
    "]\n",
    "\n",
    "with open(f'./{temp_dir}/temp_al.fna', 'w') as f:\n",
    "    SeqIO.write(records, f, 'fasta')\n",
    "    \n",
    "blast_cmd(query_primer=f'./{temp_dir}/temp_al.fna', db=db, out_path=f\"./{temp_dir}/{evi_nm}.fna\")\n",
    "blast_df = parse_blastTxt(blast_txt=f\"./{temp_dir}/{evi_nm}.fna\")\n",
    "\n",
    "info_t = {'Environment': evi_nm}\n",
    "for pri_ty, pri_atgc in zip(['forward', 'reverse'], [f_pri, r_pri]):\n",
    "    info_t[pri_ty + '_seq'] = pri_atgc\n",
    "    df_ = blast_df[blast_df['query_acc.ver'] == pri_ty].reset_index(drop=True)\n",
    "    df_.sort_values(by='evalue', inplace=True, ascending=True)\n",
    "    info_t[pri_ty + '_bitscore'] = df_.loc[0, 'bit_score']\n",
    "    if pri_ty == 'forward':\n",
    "            info_t[pri_ty + '_start'] = int(df_.loc[0, 's._start']) - (int(df_.loc[0, 'q._start']) - 1)\n",
    "    else:\n",
    "        info_t[pri_ty + '_start'] = int(df_.loc[0, 's._start']) + (int(df_.loc[0, 'q._start']) - 1)\n",
    "df_evi_primer_ecoli.append(info_t)\n",
    "\n",
    "\n",
    "df_evi_primer_ecoli = pd.DataFrame(df_evi_primer_ecoli)\n",
    "df_evi_primer_ecoli.to_csv(f'results/bacdive/evi_opt_pri_alignEcoli_{design_pri_targetVregion}.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_evi_primer_ecoli['forward_start'].hist(bins = 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_evi_primer_ecoli['reverse_start'].hist(bins = 110)\n",
    "plt.xlim([780, 830])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "notebookRunGroups": {
     "groupValue": "1"
    }
   },
   "source": [
    "#### 4. 构建用于EcoPrimer测试的demo输入数据\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "head -n 2000 /data3/hyzhang/16sDeepSeg_primer_2023/mipe_metagenome_primer_screen/16s_metagenomes_data/sputum_gq0505/metagenome/clean_ssu_seqs/zhuyoubao20190508_FDSW202112160_1r_SE1.pick.fasta > input/metagenomic_input/demo1_SE1.pick.fasta\n",
      "head -n 2000 /data3/hyzhang/16sDeepSeg_primer_2023/mipe_metagenome_primer_screen/16s_metagenomes_data/sputum_gq0505/metagenome/clean_ssu_seqs/zhaozhangxiang20190508_FDSW202112182_1r_SE1.pick.fasta > input/metagenomic_input/demo2_SE1.pick.fasta\n"
     ]
    }
   ],
   "source": [
    "demo_input_ori = pd.read_csv('input/demo_input_EcoPrimer_ori.csv')\n",
    "demo_root = 'input/metagenomic_input'\n",
    "\n",
    "for demo_i in range(demo_input_ori.shape[0]):\n",
    "    ori_path_single_end = demo_input_ori.iloc[demo_i, 2]\n",
    "    ori_path_pair_end = demo_input_ori.iloc[demo_i, 1]\n",
    "    srr_id = demo_input_ori.iloc[demo_i, 0]\n",
    "    srr_id_ori = os.path.basename(ori_path_single_end).split('_SE1.pick.fasta')[0]\n",
    "    \n",
    "    demo_single_end = os.path.join(demo_root, srr_id + '_SE1.pick.fasta')\n",
    "    demo_pair_end = os.path.join(demo_root, srr_id + '_PEmerged.pick.fasta')\n",
    "    print(f'head -n 2000 {ori_path_single_end} > {demo_single_end}')\n",
    "    os.system(f'head -n 2000 {ori_path_single_end} > {demo_single_end}')\n",
    "    os.system(f'head -n 2000 {ori_path_pair_end} > {demo_pair_end}')\n",
    "    \n",
    "    demo_input_ori.iloc[demo_i, 1] = demo_pair_end\n",
    "    demo_input_ori.iloc[demo_i, 2] = demo_single_end\n",
    "    demo_input_ori.iloc[demo_i, 3] = srr_id_ori\n",
    "\n",
    "demo_input_ori.to_csv('input/demo_input_EcoPrimer.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "demo_input_ori = pd.read_csv('input/demo_input_EcoPrimer_ori.csv')\n",
    "demo_root = 'input/metagenomic_input'\n",
    "\n",
    "for demo_i in range(demo_input_ori.shape[0]):\n",
    "    ori_path_single_end = demo_input_ori.iloc[demo_i, 2]\n",
    "    ori_path_pair_end = demo_input_ori.iloc[demo_i, 1]\n",
    "    srr_id = demo_input_ori.iloc[demo_i, 0]\n",
    "    srr_id_ori = os.path.basename(ori_path_single_end).split('_SE1.pick.fasta')[0]\n",
    "    \n",
    "    demo_single_end = os.path.join(demo_root, srr_id + '_SE1.pick.fasta')\n",
    "    demo_pair_end = os.path.join(demo_root, srr_id + '_PEmerged.pick.fasta')\n",
    "    print(f'head -n 2000 {ori_path_single_end} > {demo_single_end}')\n",
    "    os.system(f'head -n 2000 {ori_path_single_end} > {demo_single_end}')\n",
    "    os.system(f'head -n 2000 {ori_path_pair_end} > {demo_pair_end}')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "EcoPrimer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
